
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Empirical data &#8212; Dinf</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="ABC-GAN" href="abc-gan.html" />
    <link rel="prev" title="Models with multiple demes" href="multiple_demes.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Dinf</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorial
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="creating_a_model.html">
   1. Creating a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="testing_a_model.html">
   2. Testing a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="abc.html">
   3. An ABC analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Topic guides
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="accuracy.html">
   Improving discriminator accuracy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="features.html">
   Feature matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="performance.html">
   Performance characteristics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="multiple_demes.html">
   Models with multiple demes
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Empirical data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="abc-gan.html">
   ABC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mcmc-gan.html">
   MCMC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pg-gan.html">
   PG-GAN (simulated annealing)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   API reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cli.html">
   CLI reference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../development.html">
   Development
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../CHANGELOG.html">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            dinf 0.3.1.dev34
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/RacimoLab/dinf"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/RacimoLab/dinf/issues/new?title=Issue%20on%20page%20%2Fguide/empirical_data.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#vcf-bcf-files">
   VCF/BCF files
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#model-misspecification">
   Model misspecification
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#complete-example">
   Complete example
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Empirical data</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#vcf-bcf-files">
   VCF/BCF files
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#model-misspecification">
   Model misspecification
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#complete-example">
   Complete example
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="empirical-data">
<span id="sec-guide-empirical-data"></span><h1>Empirical data<a class="headerlink" href="#empirical-data" title="Permalink to this headline">#</a></h1>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>write me</p>
</div>
<section id="vcf-bcf-files">
<h2>VCF/BCF files<a class="headerlink" href="#vcf-bcf-files" title="Permalink to this headline">#</a></h2>
<p>Please use BCF rather than VCF.</p>
<p>See <a class="reference internal" href="../api.html#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-class docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>.</p>
</section>
<section id="model-misspecification">
<span id="sec-guide-misspecification"></span><h2>Model misspecification<a class="headerlink" href="#model-misspecification" title="Permalink to this headline">#</a></h2>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>move this to its own page</p>
</div>
<p>There might be differences between the features from the <code class="docutils literal notranslate"><span class="pre">target_func</span></code>
and features from the <code class="docutils literal notranslate"><span class="pre">generator_func</span></code> that are not accounted for by
the model. If this is the case, the discriminator may be able to
distinguish between the two data sources very easily.</p>
<p>Differences can be identified by first obtaining an estimate of the parameters
from the empirical data, then using these estimates for the <code class="docutils literal notranslate"><span class="pre">truth</span></code> values,
setting <code class="docutils literal notranslate"><span class="pre">target_func=None</span></code>, and training a new discriminator.
If the discriminatory power is much smaller from this simulation study
(i.e. test accuracy is much lower), this likely indicates unmodelled
differences between the empirical data and the model.</p>
<p>The following things may help to close the gap:</p>
<ul class="simple">
<li><p>Increase the model complexity. E.g. include additional
population structure in the demographic model,
or sample recombination rates from an empirically-derived
recombination map.</p></li>
<li><p>Filter the empirical data to remove genotyping errors.
E.g. remove low-coverage individuals and set a non-zero <code class="docutils literal notranslate"><span class="pre">maf_thresh</span></code>
for the feature extraction.</p></li>
<li><p>Check any other data characteristics or filtering steps,
to identify how the empirical data may be different to simulations.
E.g. SNP ascertainment.</p></li>
</ul>
</section>
<section id="complete-example">
<h2>Complete example<a class="headerlink" href="#complete-example" title="Permalink to this headline">#</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">string</span>

<span class="kn">import</span> <span class="nn">demes</span>
<span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">dinf</span>


<span class="n">populations</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;YRI&quot;</span><span class="p">,</span> <span class="s2">&quot;CEU&quot;</span><span class="p">,</span> <span class="s2">&quot;CHB&quot;</span><span class="p">]</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">get_samples_from_1kgp_metadata</span><span class="p">(</span>
    <span class="s2">&quot;20130606_g1k_3202_samples_ped_population.txt&quot;</span><span class="p">,</span> <span class="n">populations</span><span class="o">=</span><span class="n">populations</span>
<span class="p">)</span>
<span class="n">contig_lengths</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">get_contig_lengths</span><span class="p">(</span>
    <span class="s2">&quot;GRCh38_full_analysis_set_plus_decoy_hla.fa.fai&quot;</span><span class="p">,</span>
    <span class="n">keep_contigs</span><span class="o">=</span><span class="p">{</span><span class="sa">f</span><span class="s2">&quot;chr</span><span class="si">{</span><span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">21</span><span class="p">)},</span>  <span class="c1"># Exclude chrX, etc.</span>
<span class="p">)</span>
<span class="n">num_individuals</span> <span class="o">=</span> <span class="mi">64</span>
<span class="n">recombination_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>
<span class="n">mutation_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>
<span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">5_000_000</span>

<span class="n">parameters</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">Parameters</span><span class="p">(</span>
    <span class="c1"># population sizes</span>
    <span class="n">N_anc</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">30_000</span><span class="p">),</span>
    <span class="n">N_AMH</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">30_000</span><span class="p">),</span>
    <span class="n">N_OOA</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">10_000</span><span class="p">),</span>
    <span class="n">N_YRI</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">100_000</span><span class="p">),</span>
    <span class="n">N_CEU_start</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">10_000</span><span class="p">),</span>
    <span class="n">N_CEU_end</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">100_000</span><span class="p">),</span>
    <span class="n">N_CHB_start</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">10_000</span><span class="p">),</span>
    <span class="n">N_CHB_end</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">100_000</span><span class="p">),</span>
    <span class="c1"># Time units match the demography, which are specified in &quot;years&quot;.</span>
    <span class="c1"># To avoid explicitly defining constraints such as</span>
    <span class="c1">#   &quot;CEU/CHB split more recently than the OOA event&quot;,</span>
    <span class="c1"># we parameterise times as time spans, rather than absolute times.</span>
    <span class="c1"># time span of AMH</span>
    <span class="n">dT_AMH</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">200_000</span><span class="p">),</span>
    <span class="c1"># time span of OOA</span>
    <span class="n">dT_OOA</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">5_000</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">200_000</span><span class="p">),</span>
    <span class="c1"># time span of CEU and CHB.</span>
    <span class="n">dT_CEU_CHB</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">50_000</span><span class="p">),</span>
    <span class="c1"># migration rates</span>
    <span class="n">m_YRI_OOA</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">),</span>
    <span class="n">m_YRI_CEU</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">),</span>
    <span class="n">m_YRI_CHB</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">),</span>
    <span class="n">m_CEU_CHB</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">),</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">demography</span><span class="p">(</span><span class="o">**</span><span class="n">theta</span><span class="p">):</span>
    <span class="c1"># Arguments are expected to match the parameter names.</span>
    <span class="k">assert</span> <span class="n">theta</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">==</span> <span class="n">parameters</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

    <span class="n">theta</span><span class="p">[</span><span class="s2">&quot;T_OOA_end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">theta</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;dT_CEU_CHB&quot;</span><span class="p">)</span>
    <span class="n">theta</span><span class="p">[</span><span class="s2">&quot;T_AMH_end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">theta</span><span class="p">[</span><span class="s2">&quot;T_OOA_end&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">theta</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;dT_OOA&quot;</span><span class="p">)</span>
    <span class="n">theta</span><span class="p">[</span><span class="s2">&quot;T_anc_end&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">theta</span><span class="p">[</span><span class="s2">&quot;T_AMH_end&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">theta</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;dT_AMH&quot;</span><span class="p">)</span>

    <span class="n">model</span> <span class="o">=</span> <span class="n">string</span><span class="o">.</span><span class="n">Template</span><span class="p">(</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        description: The Gutenkunst et al. (2009) out-of-Africa model.</span>
<span class="sd">        doi:</span>
<span class="sd">          - https://doi.org/10.1371/journal.pgen.1000695</span>
<span class="sd">        time_units: years</span>
<span class="sd">        generation_time: 25</span>

<span class="sd">        demes:</span>
<span class="sd">          - name: ancestral</span>
<span class="sd">            epochs:</span>
<span class="sd">              - {end_time: $T_anc_end, start_size: $N_anc}</span>
<span class="sd">          - name: AMH</span>
<span class="sd">            ancestors: [ancestral]</span>
<span class="sd">            epochs:</span>
<span class="sd">              - {end_time: $T_AMH_end, start_size: $N_AMH}</span>
<span class="sd">          - name: OOA</span>
<span class="sd">            ancestors: [AMH]</span>
<span class="sd">            epochs:</span>
<span class="sd">              - {end_time: $T_OOA_end, start_size: $N_OOA}</span>
<span class="sd">          - name: YRI</span>
<span class="sd">            ancestors: [AMH]</span>
<span class="sd">            epochs:</span>
<span class="sd">              - start_size: 12300</span>
<span class="sd">          - name: CEU</span>
<span class="sd">            ancestors: [OOA]</span>
<span class="sd">            epochs:</span>
<span class="sd">              - {start_size: $N_CEU_start, end_size: $N_CEU_end}</span>
<span class="sd">          - name: CHB</span>
<span class="sd">            ancestors: [OOA]</span>
<span class="sd">            epochs:</span>
<span class="sd">              - {start_size: $N_CHB_start, end_size: $N_CHB_end}</span>

<span class="sd">        migrations:</span>
<span class="sd">          - {demes: [YRI, OOA], rate: $m_YRI_OOA}</span>
<span class="sd">          - {demes: [YRI, CEU], rate: $m_YRI_CEU}</span>
<span class="sd">          - {demes: [YRI, CHB], rate: $m_YRI_CHB}</span>
<span class="sd">          - {demes: [CEU, CHB], rate: $m_CEU_CHB}</span>
<span class="sd">        &quot;&quot;&quot;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">substitute</span><span class="p">(</span><span class="o">**</span><span class="n">theta</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">demes</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>


<span class="n">features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">MultipleBinnedHaplotypeMatrices</span><span class="p">(</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">num_individuals</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">},</span>
    <span class="n">num_loci</span><span class="o">=</span><span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="mi">128</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">},</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="mi">2</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">},</span>
    <span class="c1"># The so-called &quot;phased&quot; 1kG vcfs also contain unphased genotypes</span>
    <span class="c1"># for some individuals at some sites.</span>
    <span class="n">global_phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">global_maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">generator</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span> <span class="o">**</span><span class="n">theta</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Simulate the Gutenkunst out-of-Africa model with msprime.&quot;&quot;&quot;</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">graph</span> <span class="o">=</span> <span class="n">demography</span><span class="p">(</span><span class="o">**</span><span class="n">theta</span><span class="p">)</span>
    <span class="n">demog</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_demes</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">seed1</span><span class="p">,</span> <span class="n">seed2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">31</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">samples</span><span class="o">=</span><span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">num_individuals</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">},</span>
        <span class="n">demography</span><span class="o">=</span><span class="n">demog</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="o">=</span><span class="n">seed1</span><span class="p">,</span>
        <span class="n">record_provenance</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">seed2</span><span class="p">)</span>
    <span class="n">individuals</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">dinf</span><span class="o">.</span><span class="n">ts_individuals</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">pop</span><span class="p">)</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">}</span>
    <span class="n">labelled_matrices</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">individuals</span><span class="o">=</span><span class="n">individuals</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">labelled_matrices</span>


<span class="n">vcfs</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BagOfVcf</span><span class="p">(</span>
    <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;bcf/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*.bcf&quot;</span><span class="p">),</span>
    <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
    <span class="n">contig_lengths</span><span class="o">=</span><span class="n">contig_lengths</span><span class="p">,</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">target</span><span class="p">(</span><span class="n">seed</span><span class="p">):</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">labelled_matrices</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span>
        <span class="n">vcfs</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
        <span class="n">min_seg_sites</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
        <span class="n">max_missing_genotypes</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">labelled_matrices</span>


<span class="n">dinf_model</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">DinfModel</span><span class="p">(</span>
    <span class="n">target_func</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
    <span class="n">generator_func</span><span class="o">=</span><span class="n">generator</span><span class="p">,</span>
    <span class="n">parameters</span><span class="o">=</span><span class="n">parameters</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./guide"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="multiple_demes.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Models with multiple demes</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="abc-gan.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">ABC-GAN</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Graham Gower<br/>
  
      &copy; Copyright 2021-2022.<br/>
    <div class="extra_footer">
      dinf 0.3.1.dev34+g0efe290
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>