
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Feature matrices &#8212; Dinf</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Performance characteristics" href="performance.html" />
    <link rel="prev" title="Improving discriminator accuracy" href="accuracy.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Dinf</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorial
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="creating_a_model.html">
   1. Creating a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="testing_a_model.html">
   2. Testing a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="abc.html">
   3. An ABC analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Topic guides
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="accuracy.html">
   Improving discriminator accuracy
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Feature matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="performance.html">
   Performance characteristics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="multiple_demes.html">
   Models with multiple demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="empirical_data.html">
   Empirical data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="abc-gan.html">
   ABC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mcmc-gan.html">
   MCMC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pg-gan.html">
   PG-GAN (simulated annealing)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../api.html">
   API reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cli.html">
   CLI reference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../development.html">
   Development
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../CHANGELOG.html">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            dinf 0.3.1.dev85
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/RacimoLab/dinf"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/RacimoLab/dinf/issues/new?title=Issue%20on%20page%20%2Fguide/features.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#model">
   Model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plotting-a-feature">
   Plotting a feature
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#binned-versus-unbinned">
   Binned versus unbinned
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sequence-length">
   Sequence length
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#number-of-loci">
   Number of loci
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multiple-feature-matrices">
   Multiple feature matrices
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Feature matrices</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#model">
   Model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#plotting-a-feature">
   Plotting a feature
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#binned-versus-unbinned">
   Binned versus unbinned
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#sequence-length">
   Sequence length
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#number-of-loci">
   Number of loci
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multiple-feature-matrices">
   Multiple feature matrices
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="feature-matrices">
<span id="sec-guide-features"></span><h1>Feature matrices<a class="headerlink" href="#feature-matrices" title="Permalink to this headline">#</a></h1>
<p>This page provides a visual tour of the feature matrices provided
by Dinf.</p>
<section id="model">
<h2>Model<a class="headerlink" href="#model" title="Permalink to this headline">#</a></h2>
<p>To observe how changes to various variables will change the
resulting feature matrices, we define a <code class="docutils literal notranslate"><span class="pre">sim()</span></code> function.
This function is derived from the generator function of the bottleneck model
we’ve seen previously, and returns a feature matrix.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">demes</span>
<span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">dinf</span>


<span class="n">bottleneck_model</span> <span class="o">=</span> <span class="n">demes</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    description: Two-epoch model with recent bottleneck.</span>
<span class="sd">    time_units: generations</span>
<span class="sd">    demes:</span>
<span class="sd">      - name: A</span>
<span class="sd">        epochs:</span>
<span class="sd">          - start_size: 10000</span>
<span class="sd">            end_time: 100</span>
<span class="sd">          - start_size: 1000</span>
<span class="sd">            end_time: 0</span>
<span class="sd">    &quot;&quot;&quot;</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">sim</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="n">num_individuals</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">):</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">seed1</span><span class="p">,</span> <span class="n">seed2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">31</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">recombination_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>

    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">samples</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">demography</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_demes</span><span class="p">(</span><span class="n">bottleneck_model</span><span class="p">),</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="o">=</span><span class="n">seed1</span><span class="p">,</span>
        <span class="n">record_provenance</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">seed2</span><span class="p">)</span>

    <span class="n">feature_matrix</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">feature_matrix</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="plotting-a-feature">
<h2>Plotting a feature<a class="headerlink" href="#plotting-a-feature" title="Permalink to this headline">#</a></h2>
<p>We’ll also need some code to plot the feature matrices.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use SVG as the output format for notebooks.</span>
<span class="kn">from</span> <span class="nn">matplotlib_inline.backend_inline</span> <span class="kn">import</span> <span class="n">set_matplotlib_formats</span>
<span class="n">set_matplotlib_formats</span><span class="p">(</span><span class="s2">&quot;svg&quot;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>


<span class="k">def</span> <span class="nf">plot_feature</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">channel</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">cb_label</span><span class="o">=</span><span class="s2">&quot;Minor alleles&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Plot a feature matrix.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">channel</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;viridis&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">M</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;plasma&quot;</span>
    <span class="n">im</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span>
        <span class="n">M</span><span class="p">[</span><span class="o">...</span><span class="p">,</span> <span class="n">channel</span><span class="p">],</span>
        <span class="n">interpolation</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span>
        <span class="n">origin</span><span class="o">=</span><span class="s2">&quot;lower&quot;</span><span class="p">,</span>
        <span class="n">rasterized</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="c1"># left, right, bottom, top</span>
        <span class="n">extent</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">M</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">0</span><span class="p">,</span> <span class="n">M</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
        <span class="n">aspect</span><span class="o">=</span><span class="s2">&quot;auto&quot;</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;individuals&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;loci&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;top&quot;</span><span class="p">,</span> <span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="s2">&quot;bottom&quot;</span><span class="p">,</span> <span class="s2">&quot;left&quot;</span><span class="p">):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="n">sp</span><span class="p">]</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">cb</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">fraction</span><span class="o">=</span><span class="mf">0.04</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">cb_label</span><span class="p">)</span>
    <span class="c1"># Prefer integer ticks.</span>
    <span class="n">cb</span><span class="o">.</span><span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_major_locator</span><span class="p">(</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">ticker</span><span class="o">.</span><span class="n">MaxNLocator</span><span class="p">(</span><span class="n">integer</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">ax</span>


<span class="k">def</span> <span class="nf">get_axes</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="mi">9</span> <span class="o">/</span> <span class="mi">16</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Build subplot axes.&quot;&quot;&quot;</span>
    <span class="n">w</span><span class="p">,</span> <span class="n">h</span> <span class="o">=</span> <span class="n">scale</span> <span class="o">*</span> <span class="n">plt</span><span class="o">.</span><span class="n">figaspect</span><span class="p">(</span><span class="n">aspect</span><span class="p">)</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>
        <span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="n">ncols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">ncols</span> <span class="o">*</span> <span class="n">w</span><span class="p">,</span> <span class="n">nrows</span> <span class="o">*</span> <span class="n">h</span><span class="p">),</span> <span class="n">tight_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">axs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">axs</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axs</span><span class="o">.</span><span class="n">flat</span><span class="p">:</span>
        <span class="c1"># Get the ticklabels back for the shared axis.</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_tick_params</span><span class="p">(</span><span class="n">labelbottom</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># Prefer integer ticks.</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_major_locator</span><span class="p">(</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">ticker</span><span class="o">.</span><span class="n">MaxNLocator</span><span class="p">(</span><span class="n">integer</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">axs</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="binned-versus-unbinned">
<h2>Binned versus unbinned<a class="headerlink" href="#binned-versus-unbinned" title="Permalink to this headline">#</a></h2>
<p>Dinf supports two styles of feature matrices,
which encode information about genotypes in different ways.</p>
<p><a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a> takes the central <code class="docutils literal notranslate"><span class="pre">num_loci</span></code>
SNPs and encodes the minor allele as a <code class="docutils literal notranslate"><span class="pre">1</span></code>. For unphased data, genotypes
are summed across the chromosome copies, so matrix entries for diploid
individuals may be <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, or <code class="docutils literal notranslate"><span class="pre">2</span></code>.
If there are fewer than <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> SNPs in the genomic window, the matrix
is padded on either side with zeros.
In addition, <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a> uses two channels—effectively two
matrices—the first is the genotype matrix and the second is a matrix of
inter-SNP distances, which is repeated for each haplotype/individual.
Each inter-SNP distance is relative to the size of the genomic window,
and distances are encoded as 32 bit floats (the numpy dtype is <code class="docutils literal notranslate"><span class="pre">float32</span></code>).
Note that the two channels of the feature matrices share the dtype.</p>
<p><a class="reference internal" href="../api.html#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a> partitions the genomic window into <code class="docutils literal notranslate"><span class="pre">num_loci</span></code>
bins and counts minor alles for SNPs observed in each bin. For unphased data,
genotypes are summed across the chromosome copies.
The binning approach maintains information about the spatial distribution of
SNP along the genomic window and thus doesn’t need to encode inter-SNP
distances separately.
The magnitude of the matrix entries are driven by the density of segregating
sites relative to the size of each bin.
<a class="reference internal" href="../api.html#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a> uses 8 bit integers for the allele counts
(the numpy dtype is <code class="docutils literal notranslate"><span class="pre">int8</span></code>), which leads to a compact in-memory representation
(8 times smaller than an equivalent <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a>).</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">50_000</span>
<span class="n">num_individuals</span> <span class="o">=</span> <span class="mi">16</span>
<span class="n">hm_features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">HaplotypeMatrix</span><span class="p">(</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">num_loci</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">bhm_features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BinnedHaplotypeMatrix</span><span class="p">(</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">num_loci</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">axs</span> <span class="o">=</span> <span class="n">get_axes</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="s2">&quot;col&quot;</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="s2">&quot;row&quot;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_axis_off</span><span class="p">()</span>

<span class="n">M1</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">hm_features</span><span class="p">,</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
<span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;HaplotypeMatrix channel 0&quot;</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">channel</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cb_label</span><span class="o">=</span><span class="s2">&quot;inter-SNP distance&quot;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;HaplotypeMatrix channel 1&quot;</span><span class="p">)</span>

<span class="n">M2</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">bhm_features</span><span class="p">,</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
<span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M2</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;BinnedHaplotypeMatrix&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/features_5_0.svg" src="../_images/features_5_0.svg" /></div>
</div>
</section>
<section id="sequence-length">
<h2>Sequence length<a class="headerlink" href="#sequence-length" title="Permalink to this headline">#</a></h2>
<p>The sequence length for a feature matrix is the size of the genomic
window that was simulated or sampled from a vcf.</p>
<p>For <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a>, the sequence length should be sufficiently
long to obtain <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> SNPs <em>most of the time</em>. If the feature matrices are
being padded regularly, the sequence length should be increased
(or <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> decreased).
On the other hand, if the sequence length is too long (or <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> too small),
most diversity in the genomic window will be discarded,
which would be an inefficient use of compute resources.</p>
<p>For <a class="reference internal" href="../api.html#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a>, increasing the sequence length can
potentially extract longer range information than <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a>
(for a given <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> value).
However, the density of SNPs needs to be considered to ensure that the
allele counts per bin don’t overflow the 8 bit integer used to store the
counts. If overflow occurs regularly (i.e. for many entries in many feature
matrices), this may lead to information loss rather than information gain.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_individuals</span> <span class="o">=</span> <span class="mi">16</span>
<span class="n">hm_features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">HaplotypeMatrix</span><span class="p">(</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">num_loci</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">bhm_features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BinnedHaplotypeMatrix</span><span class="p">(</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">num_loci</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
    <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">axs</span> <span class="o">=</span> <span class="n">get_axes</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="s2">&quot;col&quot;</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="s2">&quot;row&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
    <span class="nb">zip</span><span class="p">((</span><span class="mi">5_000</span><span class="p">,</span> <span class="mi">50_000</span><span class="p">,</span> <span class="mi">500_000</span><span class="p">,</span> <span class="mi">5_000_000</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;5 kb&quot;</span><span class="p">,</span> <span class="s2">&quot;50 kb&quot;</span><span class="p">,</span> <span class="s2">&quot;500 kb&quot;</span><span class="p">,</span> <span class="s2">&quot;5 Mb&quot;</span><span class="p">))</span>
<span class="p">):</span>
    <span class="n">M1</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">features</span><span class="o">=</span><span class="n">hm_features</span><span class="p">,</span>
        <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
    <span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HaplotypeMatrix / </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">M2</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">features</span><span class="o">=</span><span class="n">bhm_features</span><span class="p">,</span>
        <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
    <span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M2</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;BinnedHaplotypeMatrix / </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/features_7_0.svg" src="../_images/features_7_0.svg" /></div>
</div>
</section>
<section id="number-of-loci">
<h2>Number of loci<a class="headerlink" href="#number-of-loci" title="Permalink to this headline">#</a></h2>
<p>Recall that the <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> feature variable indicates the number of SNPs
for a <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a>, but indicates the number of genomic bins for a
<a class="reference internal" href="../api.html#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a>.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">5_000_000</span>
<span class="n">num_individuals</span> <span class="o">=</span> <span class="mi">16</span>

<span class="n">axs</span> <span class="o">=</span> <span class="n">get_axes</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="s2">&quot;row&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">num_loci</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">((</span><span class="mi">16</span><span class="p">,</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">64</span><span class="p">,</span> <span class="mi">128</span><span class="p">,</span> <span class="mi">1024</span><span class="p">)):</span>
    <span class="n">M1</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">features</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">HaplotypeMatrix</span><span class="p">(</span>
            <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
            <span class="n">num_loci</span><span class="o">=</span><span class="n">num_loci</span><span class="p">,</span>
            <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
            <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
    <span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;HaplotypeMatrix / num_loci = </span><span class="si">{</span><span class="n">num_loci</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">M2</span> <span class="o">=</span> <span class="n">sim</span><span class="p">(</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">features</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">BinnedHaplotypeMatrix</span><span class="p">(</span>
            <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
            <span class="n">num_loci</span><span class="o">=</span><span class="n">num_loci</span><span class="p">,</span>
            <span class="n">ploidy</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
            <span class="n">phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span>
    <span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M2</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;BinnedHaplotypeMatrix / num_loci = </span><span class="si">{</span><span class="n">num_loci</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/features_9_0.svg" src="../_images/features_9_0.svg" /></div>
</div>
</section>
<section id="multiple-feature-matrices">
<h2>Multiple feature matrices<a class="headerlink" href="#multiple-feature-matrices" title="Permalink to this headline">#</a></h2>
<p>Dinf provides feature extraction classes <a class="reference internal" href="../api.html#dinf.MultipleHaplotypeMatrices" title="dinf.MultipleHaplotypeMatrices"><code class="xref py py-class docutils literal notranslate"><span class="pre">MultipleHaplotypeMatrices</span></code></a>
and <a class="reference internal" href="../api.html#dinf.MultipleBinnedHaplotypeMatrices" title="dinf.MultipleBinnedHaplotypeMatrices"><code class="xref py py-class docutils literal notranslate"><span class="pre">MultipleBinnedHaplotypeMatrices</span></code></a> that are simple extensions of
the <a class="reference internal" href="../api.html#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a> and <a class="reference internal" href="../api.html#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a> repectively.
The “Multiple” feature extraction classes produce a dictionary of features,
where each dictionary key is an arbitrary label and the value is the
feature matrix for that labelled collection of individuals.
The collections of individuals typically correspond to populations,
but this need not be the case.</p>
<p>Below, we show feature matrices for each of the demes in a two-deme
isolation-with-migration model. Each feature matrix in a given collection
may have different dimensions, i.e. different <code class="docutils literal notranslate"><span class="pre">num_loci</span></code>, and/or different
<code class="docutils literal notranslate"><span class="pre">num_individuals</span></code>. The ploidy may also differ between feature matrices,
leaving open the possibility of using “male” and “female” labels in conjunction
with sex chromosomes.
Note that MAF filtering is performed across all individuals, not within
labelled groups. Thus sites may be present in the matrices that represent
fixed differences between the labelled groups.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">im_model</span> <span class="o">=</span> <span class="n">demes</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    description: Isolation with Migration</span>
<span class="sd">    time_units: generations</span>
<span class="sd">    doi:</span>
<span class="sd">      - Wang et al. 2021, https://doi.org/10.1111/1755-0998.13386</span>
<span class="sd">    demes:</span>
<span class="sd">      - name: ancestral</span>
<span class="sd">        epochs:</span>
<span class="sd">          - start_size: 15000</span>
<span class="sd">            end_time: 2000</span>
<span class="sd">      - name: deme1</span>
<span class="sd">        ancestors: [ancestral]</span>
<span class="sd">        epochs:</span>
<span class="sd">          - start_size: 9000</span>
<span class="sd">      - name: deme2</span>
<span class="sd">        ancestors: [ancestral]</span>
<span class="sd">        epochs:</span>
<span class="sd">          - start_size: 5000</span>
<span class="sd">    pulses:</span>
<span class="sd">      - sources: [deme1]</span>
<span class="sd">        dest: deme2</span>
<span class="sd">        time: 1000</span>
<span class="sd">        proportions: [0.05]</span>
<span class="sd">    &quot;&quot;&quot;</span>
<span class="p">)</span>


<span class="k">def</span> <span class="nf">sim_im</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="n">num_individuals</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">):</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">seed1</span><span class="p">,</span> <span class="n">seed2</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">31</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">recombination_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>
    <span class="n">mutation_rate</span> <span class="o">=</span> <span class="mf">1.25e-8</span>

    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span>
        <span class="n">samples</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">demography</span><span class="o">=</span><span class="n">msprime</span><span class="o">.</span><span class="n">Demography</span><span class="o">.</span><span class="n">from_demes</span><span class="p">(</span><span class="n">im_model</span><span class="p">),</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
        <span class="n">recombination_rate</span><span class="o">=</span><span class="n">recombination_rate</span><span class="p">,</span>
        <span class="n">random_seed</span><span class="o">=</span><span class="n">seed1</span><span class="p">,</span>
        <span class="n">record_provenance</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">rate</span><span class="o">=</span><span class="n">mutation_rate</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="n">seed2</span><span class="p">)</span>

    <span class="n">populations</span> <span class="o">=</span> <span class="n">num_individuals</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="n">individuals</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">dinf</span><span class="o">.</span><span class="n">ts_individuals</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">pop</span><span class="p">)</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">}</span>
    <span class="n">labelled_features</span> <span class="o">=</span> <span class="n">features</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">individuals</span><span class="o">=</span><span class="n">individuals</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">labelled_features</span>


<span class="n">sequence_length</span> <span class="o">=</span> <span class="mi">500_000</span>
<span class="n">num_individuals</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">deme1</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">deme2</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>

<span class="n">hm_matrices</span><span class="p">,</span> <span class="n">bhm_matrices</span> <span class="o">=</span> <span class="p">(</span>
    <span class="bp">cls</span><span class="p">(</span>
        <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
        <span class="n">num_loci</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">deme1</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span> <span class="n">deme2</span><span class="o">=</span><span class="mi">48</span><span class="p">),</span>
        <span class="n">ploidy</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">deme1</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">deme2</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span>
        <span class="n">global_phased</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">global_maf_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="bp">cls</span> <span class="ow">in</span> <span class="p">(</span><span class="n">dinf</span><span class="o">.</span><span class="n">MultipleHaplotypeMatrices</span><span class="p">,</span> <span class="n">dinf</span><span class="o">.</span><span class="n">MultipleBinnedHaplotypeMatrices</span><span class="p">)</span>
<span class="p">)</span>

<span class="n">axs</span> <span class="o">=</span> <span class="n">get_axes</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="n">features1</span> <span class="o">=</span> <span class="n">sim_im</span><span class="p">(</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">hm_matrices</span><span class="p">,</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">M</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features1</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MultipleHaplotypeMatrices / </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">features2</span> <span class="o">=</span> <span class="n">sim_im</span><span class="p">(</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">bhm_matrices</span><span class="p">,</span>
    <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">,</span>
    <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">M</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">features2</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plot_feature</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;MultipleBinnedHaplotypeMatrices / </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/features_11_0.svg" src="../_images/features_11_0.svg" /></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./guide"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="accuracy.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Improving discriminator accuracy</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="performance.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Performance characteristics</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Graham Gower<br/>
  
      &copy; Copyright 2021-2022.<br/>
    <div class="extra_footer">
      dinf 0.3.1.dev85+gd5ff0ed
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>