
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>API reference &#8212; Dinf</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.1/dist/embed-amd.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="CLI reference" href="cli.html" />
    <link rel="prev" title="PG-GAN (simulated annealing)" href="guide/pg-gan.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Dinf</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorial
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="guide/creating_a_model.html">
   1. Creating a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/testing_a_model.html">
   2. Testing a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/abc.html">
   3. An ABC analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Topic guides
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="guide/accuracy.html">
   Improving discriminator accuracy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/features.html">
   Feature matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/performance.html">
   Performance characteristics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/multiple_demes.html">
   Models with multiple demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/empirical_data.html">
   Empirical data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/abc-gan.html">
   ABC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/mcmc-gan.html">
   MCMC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="guide/pg-gan.html">
   PG-GAN (simulated annealing)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   API reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cli.html">
   CLI reference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="development.html">
   Development
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="CHANGELOG.html">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            dinf 0.3.1.dev58
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/RacimoLab/dinf"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/RacimoLab/dinf/issues/new?title=Issue%20on%20page%20%2Fapi.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary">
   Summary
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#defining-a-dinf-model">
     Defining a Dinf model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#specifying-parameters">
     Specifying parameters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-extraction">
     Feature extraction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training">
     Training
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inference">
     Inference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#classification">
     Classification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#discriminator-networks">
     Discriminator networks
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#miscellaneous">
     Miscellaneous
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting">
     Plotting
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reference">
   Reference
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#type-aliases">
     Type aliases
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-defining-a-dinf-model">
     Defining a Dinf model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-parameters">
     Specifying parameters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-feature-extraction">
     Feature extraction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-train">
     Training
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-inference">
     Inference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-classification">
     Classification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-discriminator-networks">
     Discriminator networks
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-misc">
     Miscellaneous
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-plotting">
     Plotting
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>API reference</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#summary">
   Summary
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#defining-a-dinf-model">
     Defining a Dinf model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#specifying-parameters">
     Specifying parameters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#feature-extraction">
     Feature extraction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training">
     Training
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#inference">
     Inference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#classification">
     Classification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#discriminator-networks">
     Discriminator networks
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#miscellaneous">
     Miscellaneous
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting">
     Plotting
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reference">
   Reference
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#type-aliases">
     Type aliases
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-defining-a-dinf-model">
     Defining a Dinf model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-parameters">
     Specifying parameters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-feature-extraction">
     Feature extraction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-train">
     Training
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-inference">
     Inference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-classification">
     Classification
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-discriminator-networks">
     Discriminator networks
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-misc">
     Miscellaneous
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#sec-api-plotting">
     Plotting
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="api-reference">
<span id="sec-api"></span><h1>API reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">#</a></h1>
<section id="summary">
<span id="sec-api-summary"></span><h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">#</a></h2>
<section id="defining-a-dinf-model">
<h3>Defining a Dinf model<a class="headerlink" href="#defining-a-dinf-model" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DinfModel</span></code></a>(parameters, generator_func, ...[, ...])</p></td>
<td><p>A container that describes the components of a Dinf model.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="specifying-parameters">
<h3>Specifying parameters<a class="headerlink" href="#specifying-parameters" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.Parameters" title="dinf.Parameters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Parameters</span></code></a>(**kwargs)</p></td>
<td><p>An ordered collection of inferrable parameters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.Param" title="dinf.Param"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Param</span></code></a>(low, high[, truth, name])</p></td>
<td><p>An inferrable parameter with <code class="docutils literal notranslate"><span class="pre">Uniform(low,</span> <span class="pre">high)</span></code> prior distribution.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="feature-extraction">
<h3>Feature extraction<a class="headerlink" href="#feature-extraction" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a>(*, num_individuals, ...[, ...])</p></td>
<td><p>A factory for feature matrices consisting of haplotypes and relative positions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.MultipleHaplotypeMatrices" title="dinf.MultipleHaplotypeMatrices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MultipleHaplotypeMatrices</span></code></a>(*, ...[, ...])</p></td>
<td><p>A factory for labelled collections of <a class="reference internal" href="#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a> objects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a>(*, num_individuals, ...)</p></td>
<td><p>A factory for feature matrices of pseudo-haplotypes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.MultipleBinnedHaplotypeMatrices" title="dinf.MultipleBinnedHaplotypeMatrices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">MultipleBinnedHaplotypeMatrices</span></code></a>(*, ...[, ...])</p></td>
<td><p>A factory for labelled collections of <a class="reference internal" href="#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a> objects.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>(files, /, *[, samples, contig_lengths])</p></td>
<td><p>A collection of indexed VCF or BCF files.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="training">
<h3>Training<a class="headerlink" href="#training" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.train" title="dinf.train"><code class="xref py py-obj docutils literal notranslate"><span class="pre">train</span></code></a>(*, dinf_model, training_replicates, ...)</p></td>
<td><p>Train a discriminator network.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="inference">
<h3>Inference<a class="headerlink" href="#inference" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.predict" title="dinf.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>(*, discriminator, dinf_model, replicates)</p></td>
<td><p>Sample features and make predictions using the discriminator.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.abc_gan" title="dinf.abc_gan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">abc_gan</span></code></a>(*, dinf_model, iterations, ...[, ...])</p></td>
<td><p>Adversarial Abstract Bayesian Computation.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.mcmc_gan" title="dinf.mcmc_gan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mcmc_gan</span></code></a>(*, dinf_model, iterations, ...[, ...])</p></td>
<td><p>Run the MCMC GAN.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.pg_gan" title="dinf.pg_gan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pg_gan</span></code></a>(*, dinf_model, iterations, ...[, ...])</p></td>
<td><p>PG-GAN style simulated annealing.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="classification">
<h3>Classification<a class="headerlink" href="#classification" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.Discriminator" title="dinf.Discriminator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Discriminator</span></code></a>([network, state, trained, ...])</p></td>
<td><p>Wrapper for the discriminator neural network that classifies genotype matrices.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="discriminator-networks">
<h3>Discriminator networks<a class="headerlink" href="#discriminator-networks" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.ExchangeableCNN" title="dinf.ExchangeableCNN"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ExchangeableCNN</span></code></a>([sizes1, symmetric, sizes2, ...])</p></td>
<td><p>An exchangeable CNN for the discriminator.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.ExchangeablePGGAN" title="dinf.ExchangeablePGGAN"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ExchangeablePGGAN</span></code></a>([sizes1, symmetric, ...])</p></td>
<td><p>An exchangeable CNN for the discriminator using the PG-GAN architecture.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.Symmetric" title="dinf.Symmetric"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Symmetric</span></code></a>(func[, k, parent, name])</p></td>
<td><p>Network layer that summarises over a given axis in a way that is invariant to permutations of the input along that axis.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="miscellaneous">
<h3>Miscellaneous<a class="headerlink" href="#miscellaneous" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.save_results" title="dinf.save_results"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_results</span></code></a>(filename, /, *, thetas, probs, ...)</p></td>
<td><p>Save discriminator predictions and parameter values to a .npz file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.load_results" title="dinf.load_results"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_results</span></code></a>(filename, /, *[, parameters])</p></td>
<td><p>Load descriminator predictions and parameter values from a .npz file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.ts_individuals" title="dinf.ts_individuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ts_individuals</span></code></a>(ts, /[, population])</p></td>
<td><p>Get the individuals corresponding to the tree sequence's samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.get_samples_from_1kgp_metadata" title="dinf.get_samples_from_1kgp_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_samples_from_1kgp_metadata</span></code></a>(filename, /, ...)</p></td>
<td><p>Get sample IDs for 1000 Genomes Project populations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.get_contig_lengths" title="dinf.get_contig_lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_contig_lengths</span></code></a>(filename, /[, keep_contigs])</p></td>
<td><p>Load contig lengths from a whitespace-separated file like an fai (fasta index).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.Store" title="dinf.Store"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Store</span></code></a>(base, *[, create])</p></td>
<td><p>A sequence of folders with names &quot;0&quot;, &quot;1&quot;, ...</p></td>
</tr>
</tbody>
</table>
</section>
<section id="plotting">
<h3>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">#</a></h3>
<table class="autosummary longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.plot.feature" title="dinf.plot.feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">feature</span></code></a>(mat, /, *, ax[, channel, cmap, cb, ...])</p></td>
<td><p>Plot a feature matrix as a heatmap.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.plot.features" title="dinf.plot.features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">features</span></code></a>(mats, /, *[, subplots_kw])</p></td>
<td><p>Plot multiple feature matrices as heatmaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.plot.metrics" title="dinf.plot.metrics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metrics</span></code></a>(metrics_collection, /, *[, ...])</p></td>
<td><p>Plot training metrics from a discriminator neural network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#dinf.plot.hist" title="dinf.plot.hist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hist</span></code></a>(x, /, *[, ax, ci, truth, hist_kw])</p></td>
<td><p>Plot a histogram.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#dinf.plot.hist2d" title="dinf.plot.hist2d"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hist2d</span></code></a>(x, y, /, *[, x_label, y_label, ...])</p></td>
<td><p>Plot a 2d histogram.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">#</a></h2>
<section id="type-aliases">
<span id="sec-api-type-aliases"></span><h3>Type aliases<a class="headerlink" href="#type-aliases" title="Permalink to this headline">#</a></h3>
<dl class="py data">
<dt class="sig sig-object py" id="dinf.FeatureCollection">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">FeatureCollection</span></span><a class="headerlink" href="#dinf.FeatureCollection" title="Permalink to this definition">#</a></dt>
<dd><p>A feature array or a labelled collection of feature arrays.</p>
<p>alias of <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>]]</p>
</dd></dl>

</section>
<section id="sec-api-defining-a-dinf-model">
<span id="id1"></span><h3>Defining a Dinf model<a class="headerlink" href="#sec-api-defining-a-dinf-model" title="Permalink to this headline">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="dinf.DinfModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">DinfModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generator_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">discriminator_network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf_model.html#DinfModel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.DinfModel" title="Permalink to this definition">#</a></dt>
<dd><p>A container that describes the components of a Dinf model.</p>
<p>Constructing a Dinf model requires specifying:</p>
<blockquote>
<div><ul class="simple">
<li><p>the inferrable <code class="docutils literal notranslate"><span class="pre">parameters</span></code>,</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">generator_func</span></code> function that accepts concrete parameter values,
produces data under some simulation model, and returns a feature matrix
(or matrices), and</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">target_func</span></code> function that samples from the target dataset,
and returns a feature matrix (or matrices).</p></li>
</ul>
</div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dinf</span>
<span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">parameters</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">Parameters</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">features</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BinnedHaplotypeMatrix</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">vcfs</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BagOfVcf</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">generator</span><span class="p">(</span><span class="n">seed</span><span class="p">,</span> <span class="o">...</span><span class="p">):</span>
    <span class="o">...</span>
    <span class="c1"># E.g. simulate using msprime to get a succinct tree sequence (ts).</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_ancestry</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
    <span class="n">mts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">sim_mutations</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">features</span><span class="o">.</span><span class="n">from_ts</span><span class="p">(</span><span class="n">mts</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">target</span><span class="p">(</span><span class="n">seed</span><span class="p">):</span>
    <span class="o">...</span>
    <span class="k">return</span> <span class="n">features</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="n">vcfs</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>

<span class="n">dinf_model</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">DinfModel</span><span class="p">(</span>
    <span class="n">parameters</span><span class="o">=</span><span class="n">parameters</span><span class="p">,</span>
    <span class="n">generator_func</span><span class="o">=</span><span class="n">generator</span><span class="p">,</span>
    <span class="n">target_func</span><span class="o">=</span><span class="n">target</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameters</strong> (<a class="reference internal" href="#dinf.Parameters" title="dinf.Parameters"><em>Parameters</em></a>) – A collection of inferrable parameters, one for each keyword argument
provided to the <a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-attr docutils literal notranslate"><span class="pre">generator_func</span></code></a>.</p></li>
<li><p><strong>generator_func</strong> (<em>Callable</em><em>[</em><em>...</em><em>, </em><em>FeatureCollection</em><em>]</em>) – <p>A function that returns features for concrete parameter values.</p>
<p>The first (positional) argument to the function is a
<a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><code class="xref py py-class docutils literal notranslate"><span class="pre">seed</span></code></a>,
that may be used to seed a random number generator.
The subsequent (keyword) arguments correspond to concrete values for
the <a class="reference internal" href="#dinf.DinfModel.parameters" title="dinf.DinfModel.parameters"><code class="xref py py-attr docutils literal notranslate"><span class="pre">parameters</span></code></a>.
The return type is either:</p>
<blockquote>
<div><ul>
<li><p>a feature matrix, i.e. an n-dimensional numpy array, or</p></li>
<li><p>multiple feature matrices, i.e. a dictionary of n-dimensional
numpy arrays, where the dictionary keys are arbitrary labels
such as population names.</p></li>
</ul>
</div></blockquote>
<p>The signature of the function will follow the patterns below,
where <code class="docutils literal notranslate"><span class="pre">p0</span></code>, <code class="docutils literal notranslate"><span class="pre">p1</span></code>, etc. are the names of the parameters in
<a class="reference internal" href="#dinf.DinfModel.parameters" title="dinf.DinfModel.parameters"><code class="xref py py-attr docutils literal notranslate"><span class="pre">parameters</span></code></a>.
Type annotations are provided here for clarity, but are not required
in user-defined functions. In function signatures,
<a class="reference external" href="https://peps.python.org/pep-0570/">positional-only parameters</a>
preceed a <code class="docutils literal notranslate"><span class="pre">/</span></code> and
<a class="reference external" href="https://peps.python.org/pep-3102/">keyword-only parameters</a>
follow a <code class="docutils literal notranslate"><span class="pre">*</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dinf</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">parameters</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">Parameters</span><span class="p">(</span>
    <span class="n">p0</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="o">...</span><span class="p">),</span>
    <span class="n">p1</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="o">...</span><span class="p">),</span>
    <span class="o">...</span>
<span class="p">)</span>

<span class="c1"># Signature for returning a single feature matrix.</span>
<span class="k">def</span> <span class="nf">generator_func1</span><span class="p">(</span>
    <span class="n">seed</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">SeedSequence</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">p0</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">p1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="o">...</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># Signature for returning multiple feature matrices.</span>
<span class="k">def</span> <span class="nf">generator_func2</span><span class="p">(</span>
    <span class="n">seed</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">SeedSequence</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">p0</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">p1</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="o">...</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
    <span class="o">...</span>

<span class="c1"># For generator functions accepting large numbers of parameters,</span>
<span class="c1"># the following pattern using ``**kwargs`` may be preferred.</span>
<span class="k">def</span> <span class="nf">generator_func3</span><span class="p">(</span>
    <span class="n">seed</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">SeedSequence</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">float</span>
<span class="p">)</span><span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
    <span class="k">assert</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">==</span> <span class="n">parameters</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="c1"># do something with p0</span>
    <span class="n">do_something</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;p0&quot;</span><span class="p">])</span>
    <span class="c1"># do something with p1</span>
    <span class="n">do_something_else</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;p1&quot;</span><span class="p">])</span>
    <span class="o">...</span>
</pre></div>
</div>
</p></li>
<li><p><strong>target_func</strong> (<em>Callable</em><em>[</em><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><em>numpy.random.SeedSequence</em></a><em>]</em><em>, </em><em>FeatureCollection</em><em>] </em><em>| </em><em>None</em>) – <p>A function that returns features sampled from the target dataset.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">None</span></code> (which must be specified explicitly),
the <a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-attr docutils literal notranslate"><span class="pre">generator_func</span></code></a> will be used to simulate
the target dataset using each parameter’s <a class="reference internal" href="#dinf.Param.truth" title="dinf.Param.truth"><code class="xref py py-attr docutils literal notranslate"><span class="pre">truth</span></code></a>
value.</p>
<p>The function takes a single (positional) argument, a
<a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><code class="xref py py-class docutils literal notranslate"><span class="pre">seed</span></code></a>,
that may be used to seed a random number generator.
The return type must match the return type of <a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-attr docutils literal notranslate"><span class="pre">generator_func</span></code></a>.</p>
</p></li>
<li><p><strong>discriminator_network</strong> (<em>nn.Module</em><em> | </em><em>None</em>) – A <a class="reference external" href="https://flax.readthedocs.io/en/latest/index.html" title="(in Flax)"><span class="xref std std-doc">flax</span></a> neural network.
If not specified, <a class="reference internal" href="#dinf.ExchangeableCNN" title="dinf.ExchangeableCNN"><code class="xref py py-class docutils literal notranslate"><span class="pre">ExchangeableCNN</span></code></a> will be used.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.parameters">
<span class="sig-name descname"><span class="pre">parameters</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#dinf.Parameters" title="dinf.Parameters"><span class="pre">Parameters</span></a></em><a class="headerlink" href="#dinf.DinfModel.parameters" title="Permalink to this definition">#</a></dt>
<dd><p>The inferrable parameters.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.generator_func">
<span class="sig-name descname"><span class="pre">generator_func</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Callable</span><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">FeatureCollection</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.DinfModel.generator_func" title="Permalink to this definition">#</a></dt>
<dd><p>Function that simulates features using concrete parameter values.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.generator_func_v">
<span class="sig-name descname"><span class="pre">generator_func_v</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Callable</span><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><span class="pre">numpy.random.SeedSequence</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><span class="pre">float</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">FeatureCollection</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.DinfModel.generator_func_v" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper for <code class="docutils literal notranslate"><span class="pre">generator_func</span></code> that accepts a single argument containing
the seed and an iterable of parameter values (as opposed to keyword arguments).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.target_func">
<span class="sig-name descname"><span class="pre">target_func</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Callable</span><span class="p"><span class="pre">[</span></span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><span class="pre">numpy.random.SeedSequence</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">FeatureCollection</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><a class="headerlink" href="#dinf.DinfModel.target_func" title="Permalink to this definition">#</a></dt>
<dd><p>Function that samples features from the target distribution.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.discriminator_network">
<span class="sig-name descname"><span class="pre">discriminator_network</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">nn.Module</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#dinf.DinfModel.discriminator_network" title="Permalink to this definition">#</a></dt>
<dd><p>A <a class="reference external" href="https://flax.readthedocs.io/en/latest/index.html" title="(in Flax)"><span class="xref std std-doc">flax</span></a> neural network. May be <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.DinfModel.filename">
<span class="sig-name descname"><span class="pre">filename</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><span class="pre">pathlib.Path</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#dinf.DinfModel.filename" title="Permalink to this definition">#</a></dt>
<dd><p>Path to the file from which the model was loaded (if any).
May be <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.DinfModel.check">
<span class="sig-name descname"><span class="pre">check</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf_model.html#DinfModel.check"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.DinfModel.check" title="Permalink to this definition">#</a></dt>
<dd><p>Basic health checks: draw parameters and call the functions.</p>
<p>We don’t do this when initialising the DinfModel, because calling
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-attr docutils literal notranslate"><span class="pre">generator_func()</span></code></a> and <a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-attr docutils literal notranslate"><span class="pre">target_func()</span></code></a> are potentially
time consuming, which could lead to annoying delays for the command
line interface.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.DinfModel.from_file">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf_model.html#DinfModel.from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.DinfModel.from_file" title="Permalink to this definition">#</a></dt>
<dd><p>Load the symbol “dinf_model” from a file.</p>
<p><a class="reference external" href="https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly">https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – </p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel">DinfModel</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="sec-api-parameters">
<span id="id2"></span><h3>Specifying parameters<a class="headerlink" href="#sec-api-parameters" title="Permalink to this headline">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="dinf.Parameters">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">Parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters" title="Permalink to this definition">#</a></dt>
<dd><p>An ordered collection of inferrable parameters.</p>
<p>A collection of parameters is built by passing <a class="reference internal" href="#dinf.Param" title="dinf.Param"><code class="xref py py-class docutils literal notranslate"><span class="pre">Param</span></code></a> objects
to the constructor by keyword. The keyword will be used as the
<a class="reference internal" href="#dinf.Param.name" title="dinf.Param.name"><code class="xref py py-attr docutils literal notranslate"><span class="pre">name</span></code></a> for the given parameter. This class implements
Python’s <a class="reference external" href="https://docs.python.org/3/glossary.html#term-mapping" title="(in Python v3.10)"><span>mapping</span></a> protocol.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dinf</span>

<span class="c1"># Parameters named &quot;a&quot;, &quot;b&quot;, and &quot;c&quot;.</span>
<span class="n">parameters</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">Parameters</span><span class="p">(</span>
    <span class="n">a</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
    <span class="n">b</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
    <span class="n">c</span><span class="o">=</span><span class="n">dinf</span><span class="o">.</span><span class="n">Param</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">),</span>
<span class="p">)</span>

<span class="c1"># Iterate over parameters.</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span>
<span class="k">assert</span> <span class="nb">list</span><span class="p">(</span><span class="n">parameters</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">]</span>
<span class="k">assert</span> <span class="p">[</span><span class="n">p</span><span class="o">.</span><span class="n">high</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">parameters</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mf">1e-3</span><span class="p">]</span>

<span class="c1"># Lookup a parameter by name.</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">parameters</span><span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">]</span>
<span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">low</span> <span class="o">==</span> <span class="mi">0</span>
<span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">high</span> <span class="o">==</span> <span class="mi">1</span>
<span class="k">assert</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;a&quot;</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>kwargs</strong> (<a class="reference internal" href="#dinf.Param" title="dinf.Param"><em>Param</em></a>) – The <a class="reference internal" href="#dinf.Param" title="dinf.Param"><code class="xref py py-class docutils literal notranslate"><span class="pre">Param</span></code></a> objects, named by keyword.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.bounds_contain">
<span class="sig-name descname"><span class="pre">bounds_contain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.bounds_contain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.bounds_contain" title="Permalink to this definition">#</a></dt>
<dd><p>Test if values are within in the parameter bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – A 2d numpy array, where thetas[j][k] is the j’th sample for the
k’th parameter.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 1d numpy array of boolean values, where ret[j] is True if the
j’th sample is within the parameter space, and False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.sample_prior">
<span class="sig-name descname"><span class="pre">sample_prior</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.sample_prior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.sample_prior" title="Permalink to this definition">#</a></dt>
<dd><p>Get a random sample of parameter values from their prior distributions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The sample size.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – The numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 2d numpy array of parameter values, where ret[j][k] is the
j’th draw for the k’th parameter.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.sample_kde">
<span class="sig-name descname"><span class="pre">sample_kde</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.sample_kde"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.sample_kde" title="Permalink to this definition">#</a></dt>
<dd><p>Sample from a smoothed set of weighted observations.</p>
<p>Samples are drawn from <code class="docutils literal notranslate"><span class="pre">thetas</span></code>, weighted by their probability.
New points are drawn within a neighbourhood of the sampled thetas
using a mulivariate normal whose covariance is calculated from the
thetas. This is equivalent to sampling from a Gaussian KDE, but
avoids doing an explicit density estimation.
Values are sampled until they are within the parameter bounds.
Scott’s rule of thumb is used for bandwidth selection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – Parameter values to sample from.</p></li>
<li><p><strong>probs</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – Discriminator predictions corresponding to the <code class="docutils literal notranslate"><span class="pre">thetas</span></code>.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of samples to draw.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sampled values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.sample_ball">
<span class="sig-name descname"><span class="pre">sample_ball</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">theta</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.sample_ball"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.sample_ball" title="Permalink to this definition">#</a></dt>
<dd><p>Sample a ball around the given <code class="docutils literal notranslate"><span class="pre">theta</span></code> value.</p>
<p>The size of the ball is controlled by the covariance matrix, <code class="docutils literal notranslate"><span class="pre">cov</span></code>,
which is multiplied by <code class="docutils literal notranslate"><span class="pre">cov_factor</span></code>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Samples are not gauranteed to be within the parameter bounds.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>theta</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – Parameter value around which the ball is centered.</p></li>
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of samples to draw.</p></li>
<li><p><strong>cov</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><em>Optional</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a><em>]</em>) – The covariance matrix.
If not specified, a diagonal matrix will be used that has
entry cov[j][j] set to the squared range of parameter j:
(high - low)**2.</p></li>
<li><p><strong>cov_factor</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – Left-multiply the covariance matrix by this value.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sampled values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.transform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.transform" title="Permalink to this definition">#</a></dt>
<dd><p>Transform values bounded by [param.low, param.high] to [-inf, inf].</p>
<p>See <a class="reference internal" href="#dinf.Parameters.itransform" title="dinf.Parameters.itransform"><code class="xref py py-meth docutils literal notranslate"><span class="pre">itransform()</span></code></a> for the inverse transformation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be transformed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transformed parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.itransform">
<span class="sig-name descname"><span class="pre">itransform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.itransform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.itransform" title="Permalink to this definition">#</a></dt>
<dd><p>Transform values on [-inf, inf] to be bounded by [param.low, param.high].</p>
<p>Performs the inverse of <a class="reference internal" href="#dinf.Parameters.transform" title="dinf.Parameters.transform"><code class="xref py py-meth docutils literal notranslate"><span class="pre">transform()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be transformed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transformed parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.truncate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.truncate" title="Permalink to this definition">#</a></dt>
<dd><p>Truncate values that are out of bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be truncated.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Truncated parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.reflect">
<span class="sig-name descname"><span class="pre">reflect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.reflect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.reflect" title="Permalink to this definition">#</a></dt>
<dd><p>Reflect values that are out of bounds by the amount they are out.</p>
<p>As reflecting does not gaurantee values will be within the bounds,
values are first truncated to (2*low - high, 2*high - low),
then reflected. For example, with bounds low=0, high=10,
a value of -11 will be truncated to -10, then reflected to attain
a final value of 10.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be reflected.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Reflected parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.geometric_median">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">geometric_median</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.geometric_median"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.geometric_median" title="Permalink to this definition">#</a></dt>
<dd><p>Get the multivariate median of a weighted sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thetas</strong> (<em>np.ndarray</em>) – Parameter values. thetas[j][k] is the value of the k’th parameter
for the j’th multivariate sample.</p></li>
<li><p><strong>probs</strong> (<em>np.ndarray</em><em> | </em><em>None</em>) – Discriminator predictions corresponding to the <code class="docutils literal notranslate"><span class="pre">thetas</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Median position in multivariate space.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Parameters.top_n">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">top_n</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Parameters.top_n"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Parameters.top_n" title="Permalink to this definition">#</a></dt>
<dd><p>Get the top <code class="docutils literal notranslate"><span class="pre">n</span></code> values of <code class="docutils literal notranslate"><span class="pre">thetas</span></code>, ranked by <code class="docutils literal notranslate"><span class="pre">probs</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>thetas</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – Parameter values. thetas[j][k] is the value of the k’th parameter
for the j’th multivariate sample.</p></li>
<li><p><strong>probs</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – Discriminator predictions corresponding to the <code class="docutils literal notranslate"><span class="pre">thetas</span></code>.</p></li>
<li><p><strong>n</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 2-tuple of (top_thetas, top_probs).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.Param">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">Param</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">low</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">high</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param" title="Permalink to this definition">#</a></dt>
<dd><p>An inferrable parameter with <code class="docutils literal notranslate"><span class="pre">Uniform(low,</span> <span class="pre">high)</span></code> prior distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>low</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – The lower bound for the parameter.</p></li>
<li><p><strong>high</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – The uppper bound for the parameter.</p></li>
<li><p><strong>truth</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – The true value of the parameter. Set to None if not known
(e.g. for parameters corresponding to an empirical dataset).
If the value is not None, this value may be used as the
truth in a simulation study.</p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.Param.low">
<span class="sig-name descname"><span class="pre">low</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><span class="pre">float</span></a></em><a class="headerlink" href="#dinf.Param.low" title="Permalink to this definition">#</a></dt>
<dd><p>Parameter’s lower bound.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.Param.high">
<span class="sig-name descname"><span class="pre">high</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><span class="pre">float</span></a></em><a class="headerlink" href="#dinf.Param.high" title="Permalink to this definition">#</a></dt>
<dd><p>Parameter’s upper bound.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.Param.truth">
<span class="sig-name descname"><span class="pre">truth</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><span class="pre">float</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#dinf.Param.truth" title="Permalink to this definition">#</a></dt>
<dd><p>The true value of the parameter. May be None.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.Param.name">
<span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#dinf.Param.name" title="Permalink to this definition">#</a></dt>
<dd><p>Name of the parameter.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.sample_prior">
<span class="sig-name descname"><span class="pre">sample_prior</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.sample_prior"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.sample_prior" title="Permalink to this definition">#</a></dt>
<dd><p>Get a random sample from the <code class="docutils literal notranslate"><span class="pre">Uniform(low,</span> <span class="pre">high)</span></code> prior distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The sample size.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – The numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A numpy array of parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.bounds_contain">
<span class="sig-name descname"><span class="pre">bounds_contain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.bounds_contain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.bounds_contain" title="Permalink to this definition">#</a></dt>
<dd><p>Test if values are within in the parameter bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values that will be bounds checked.
May be a single number or an array.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A boolean value, or array of booleans with the same length as
the input. ret[j] is True if x[j] is in bounds, False otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.transform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.transform" title="Permalink to this definition">#</a></dt>
<dd><p>Transform bounded values on [low, high] to unbounded values on [-inf, inf].</p>
<p>Performs a logit transformation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be transformed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transformed parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.itransform">
<span class="sig-name descname"><span class="pre">itransform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.itransform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.itransform" title="Permalink to this definition">#</a></dt>
<dd><p>Transform unbounded values on [-inf, inf] to bounded values on [low, high].</p>
<p>Performs an inverse logit transformation (aka expit, aka sigmoid).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be transformed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transformed parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.truncate">
<span class="sig-name descname"><span class="pre">truncate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.truncate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.truncate" title="Permalink to this definition">#</a></dt>
<dd><p>Truncate values that are out of bounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be truncated.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Truncated parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Param.reflect">
<span class="sig-name descname"><span class="pre">reflect</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/parameters.html#Param.reflect"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Param.reflect" title="Permalink to this definition">#</a></dt>
<dd><p>Reflect values that are out of bounds by the amount they are out.</p>
<p>As reflecting does not gaurantee values will be within the bounds,
values are first truncated to (2*low - high, 2*high - low),
then reflected. For example, with bounds low=0, high=10,
a value of -11 will be truncated to -10, then reflected to attain
a final value of 10.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>x</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a>) – The values to be reflected.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Reflected parameter values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="sec-api-feature-extraction">
<span id="id3"></span><h3>Feature extraction<a class="headerlink" href="#sec-api-feature-extraction" title="Permalink to this headline">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="dinf.HaplotypeMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">HaplotypeMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_individuals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phased</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/feature_extractor.html#HaplotypeMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.HaplotypeMatrix" title="Permalink to this definition">#</a></dt>
<dd><p>A factory for feature matrices consisting of haplotypes and relative positions.</p>
<p>The feature is an <span class="math notranslate nohighlight">\(n \times m \times c\)</span> array, where the channel
dimension <span class="math notranslate nohighlight">\(c\)</span> is 2. The first channel is a haplotype matrix and
the second channel is a matrix of relative SNP positions.</p>
<p>The haplotype matrix is an <span class="math notranslate nohighlight">\(n \times m\)</span> matrix,
where <span class="math notranslate nohighlight">\(n\)</span> corresponds to the number of haplotypes
(or number of individuals, for unphased data) and <span class="math notranslate nohighlight">\(m\)</span> corresponds to
the number of SNPs along the sequence.
For phased data, each entry is a 0 or 1 corresponding to
the major or minor allele respectively. For unphased data, each entry
is the count of minor alleles across all chromosomes in the individual.
Only polymorphic SNPs are considered, and for multiallelic sites,
only the first two alleles are used.
Alleles are polarised by choosing the most frequent allele to be encoded
as 0 (the major allele), and the second most frequent allele as 1
(the minor allele).</p>
<p>The position matrix is an <span class="math notranslate nohighlight">\(n \times m\)</span> matrix, where the vector
of <span class="math notranslate nohighlight">\(m\)</span> inter-SNP distances are repeated <span class="math notranslate nohighlight">\(n\)</span> times—once
for each haplotype (or each individual, for unphased data). Each entry
is the distance from the previous SNP (as a proportion of the sequence
length). The first inter-SNP distance in the vector is always zero.</p>
<div class="line-block">
<div class="line">Chan et al. 2018, <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7687905/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7687905/</a></div>
<div class="line">Wang et al. 2021, <a class="reference external" href="https://doi.org/10.1111/1755-0998.13386">https://doi.org/10.1111/1755-0998.13386</a></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The number of individuals to include in the feature matrix.</p></li>
<li><p><strong>num_loci</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The number of SNP sites to extract.
The central <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> SNPs in the sequence will be used.
If there are fewer than <code class="docutils literal notranslate"><span class="pre">num_loci</span></code> SNPs, the feature matrix will
be padded on both sides with zeros.</p></li>
<li><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Ploidy of the individuals.</p></li>
<li><p><strong>maf_thresh</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Minor allele frequency (MAF) threshold. Sites with MAF lower than
this value are ignored. If None, only invariant sites will be excluded.</p></li>
<li><p><strong>phased</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, the individuals’ haplotypes will each be included as
independent rows in the feature matrix and the shape of the
feature matrix will be <code class="docutils literal notranslate"><span class="pre">(ploidy</span> <span class="pre">*</span> <span class="pre">num_individuals,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.
If False, the allele counts for each individual will be summed
across their chromosome copies and the shape of the feature matrix
will be <code class="docutils literal notranslate"><span class="pre">(num_individuals,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.HaplotypeMatrix.from_ts">
<span class="sig-name descname"><span class="pre">from_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.HaplotypeMatrix.from_ts" title="Permalink to this definition">#</a></dt>
<dd><p>Create a feature matrix from a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/introduction.html#sec-introduction" title="(in Python)"><span class="xref std std-ref">tskit</span></a>
tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><em>tskit.trees.TreeSequence</em></a>) – The tree sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.HaplotypeMatrix.from_vcf">
<span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vb</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_missing_genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_seg_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.HaplotypeMatrix.from_vcf" title="Permalink to this definition">#</a></dt>
<dd><p>Create a feature matrix from a region of a VCF/BCF.</p>
<p>The genomic window for a feature matrix is drawn uniformly at random
from the contigs defined in the given <a class="reference internal" href="#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-class docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>, <code class="docutils literal notranslate"><span class="pre">vb</span></code>.</p>
<p>Individuals in the VCFs are sampled (without replacement) for
inclusion in the output matrix. The size of the feature space
can therefore be vastly increased by having more individuals
in the VCFs than are needed for the feature dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vb</strong> (<a class="reference internal" href="#dinf.BagOfVcf" title="dinf.vcf.BagOfVcf"><em>dinf.vcf.BagOfVcf</em></a>) – The BagOfVcf object that describes the VCF/BCF files.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the genomic window to be sampled.</p></li>
<li><p><strong>max_missing_genotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Consider only sites with at most this many missing genotype calls.</p></li>
<li><p><strong>min_seg_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Sampled genotype matrix must have at least this many variable
sites (after filtering sites for missingness).</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="dinf.HaplotypeMatrix.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.HaplotypeMatrix.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Shape of the feature matrix.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.MultipleHaplotypeMatrices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">MultipleHaplotypeMatrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_individuals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_maf_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_phased</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/feature_extractor.html#MultipleHaplotypeMatrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.MultipleHaplotypeMatrices" title="Permalink to this definition">#</a></dt>
<dd><p>A factory for labelled collections of <a class="reference internal" href="#dinf.HaplotypeMatrix" title="dinf.HaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">HaplotypeMatrix</span></code></a> objects.</p>
<p>One feature matrix is produced for each label. Labels correspond to
collections of individuals that will be treated as exchangeable
(e.g. populations).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_individuals</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the number of individuals
in the feature matrix.</p></li>
<li><p><strong>num_loci</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the number of feature loci to be
extracted from the sequence.</p></li>
<li><p><strong>ploidy</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the ploidy of the individuals.</p></li>
<li><p><strong>global_maf_thresh</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Minor allele frequency (MAF) threshold. Sites with MAF lower than
this value are ignored. MAF is calculated across all individuals.
If None, only invariant sites will be excluded.</p></li>
<li><p><strong>global_phased</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, the individuals’ haplotypes will each be included as
independent rows in each feature matrix and the shape of the
feature matrix for label <code class="docutils literal notranslate"><span class="pre">l</span></code> will be
<code class="docutils literal notranslate"><span class="pre">(ploidy[l]</span> <span class="pre">*</span> <span class="pre">num_individuals[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
If False, the allele counts for each individual will be summed
across their chromosome copies and the shape of the feature matrix
will be <code class="docutils literal notranslate"><span class="pre">(num_individuals[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.MultipleHaplotypeMatrices.from_ts">
<span class="sig-name descname"><span class="pre">from_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.MultipleHaplotypeMatrices.from_ts" title="Permalink to this definition">#</a></dt>
<dd><p>Create pseudo-genotype matrices from a tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><em>tskit.trees.TreeSequence</em></a>) – The tree sequence.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><em>Mapping</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.10)"><em>Any</em></a><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v1.23)"><em>numpy.dtype</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.integer" title="(in NumPy v1.23)"><em>numpy.integer</em></a><em>]</em><em>]</em><em>]</em>) – A mapping from label to an array of individuals.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary mapping a label <code class="docutils literal notranslate"><span class="pre">l</span></code> to a feature array.
Each array has shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
For an array <span class="math notranslate nohighlight">\(M\)</span>, the <span class="math notranslate nohighlight">\(M[i][j][0]\)</span>’th entry is the
count of minor alleles in the <span class="math notranslate nohighlight">\(j\)</span>’th bin of psdeudo-haplotype
<span class="math notranslate nohighlight">\(i\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.10)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.MultipleHaplotypeMatrices.from_vcf">
<span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vb</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_missing_genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_seg_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.MultipleHaplotypeMatrices.from_vcf" title="Permalink to this definition">#</a></dt>
<dd><p>Create pseudo-genotype matrices from a region of a VCF/BCF.</p>
<p>The genomic window is drawn uniformly at random from the sequences
defined in the given <a class="reference internal" href="#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-class docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>.</p>
<p>Individuals in the VCFs are sampled (without replacement) for
inclusion in the output matrix. The size of the feature space
can therefore be vastly increased by having more individuals
in the VCFs than are needed for the feature dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vb</strong> (<a class="reference internal" href="#dinf.BagOfVcf" title="dinf.vcf.BagOfVcf"><em>dinf.vcf.BagOfVcf</em></a>) – A collection of indexed VCF/BCF files.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the genomic window to be sampled.</p></li>
<li><p><strong>max_missing_genotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Consider only sites with fewer missing genotype calls than
this number.</p></li>
<li><p><strong>min_seg_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Sampled genotype matrix must have at least this many variable
sites (after filtering sites for missingness).</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary mapping a label <code class="docutils literal notranslate"><span class="pre">l</span></code> to a feature array.
Each array has shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
For an array <span class="math notranslate nohighlight">\(M\)</span>, the <span class="math notranslate nohighlight">\(M[i][j][0]\)</span>’th entry is the
count of minor alleles in the <span class="math notranslate nohighlight">\(j\)</span>’th bin of psdeudo-haplotype
<span class="math notranslate nohighlight">\(i\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.10)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="dinf.MultipleHaplotypeMatrices.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.MultipleHaplotypeMatrices.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Shape of the feature matrices.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.BinnedHaplotypeMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">BinnedHaplotypeMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_individuals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phased</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maf_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/feature_extractor.html#BinnedHaplotypeMatrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.BinnedHaplotypeMatrix" title="Permalink to this definition">#</a></dt>
<dd><p>A factory for feature matrices of pseudo-haplotypes.</p>
<p>The binned haplotype matrix is an <span class="math notranslate nohighlight">\(n \times m\)</span> matrix,
where <span class="math notranslate nohighlight">\(n\)</span> corresponds to the number of haplotypes
(or polyploid genotypes, for unphased data) and <span class="math notranslate nohighlight">\(m\)</span> corresponds to
a set of equally sized bins along the sequence length. Each matrix entry
contains the count of minor alleles in an individual’s haplotype
(or polyploid genotype) in a given bin.</p>
<p>Only polymorphic SNPs are considered, and for multiallelic sites,
only the first two alleles are used.
Alleles are polarised by choosing the most frequent allele to be encoded
as 0, and the second most frequent allele as 1.</p>
<p>As the features are intended to be passed to a covolutional neural network,
the output dimensions are actually <span class="math notranslate nohighlight">\(n \times m \times 1\)</span>, where the
final dimension is the (unused) “channels” dimension for the convolution.</p>
<p>Gower et al. 2021, <a class="reference external" href="https://doi.org/10.7554/eLife.64669">https://doi.org/10.7554/eLife.64669</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The number of individuals to include in the feature matrix.</p></li>
<li><p><strong>num_loci</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The number of bins into which the sequence is partitioned.
Each bin spans <code class="docutils literal notranslate"><span class="pre">sequence_length</span> <span class="pre">/</span> <span class="pre">num_loci</span></code> base pairs.</p></li>
<li><p><strong>ploidy</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Ploidy of the individuals.</p></li>
<li><p><strong>phased</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, the individuals’ haplotypes will each be included as
independent rows in the feature matrix and the shape of the
feature matrix will be <code class="docutils literal notranslate"><span class="pre">(ploidy</span> <span class="pre">*</span> <span class="pre">num_individuals,</span> <span class="pre">num_loci,</span> <span class="pre">1)</span></code>.
If False, the allele counts for each individual will be summed
across their chromosome copies and the shape of the feature matrix
will be <code class="docutils literal notranslate"><span class="pre">(num_individuals,</span> <span class="pre">num_loci,</span> <span class="pre">1)</span></code>.</p></li>
<li><p><strong>maf_thresh</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Minor allele frequency (MAF) threshold. Sites with MAF lower than
this value are ignored. If None, only invariant sites will be excluded.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.BinnedHaplotypeMatrix.from_ts">
<span class="sig-name descname"><span class="pre">from_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.BinnedHaplotypeMatrix.from_ts" title="Permalink to this definition">#</a></dt>
<dd><p>Create a feature matrix from a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/introduction.html#sec-introduction" title="(in Python)"><span class="xref std std-ref">tskit</span></a>
tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><em>tskit.trees.TreeSequence</em></a>) – The tree sequence.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.BinnedHaplotypeMatrix.from_vcf">
<span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vb</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_missing_genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_seg_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.BinnedHaplotypeMatrix.from_vcf" title="Permalink to this definition">#</a></dt>
<dd><p>Create a feature matrix from a region of a VCF/BCF.</p>
<p>The genomic window for a feature matrix is drawn uniformly at random
from the contigs defined in the given <a class="reference internal" href="#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-class docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>, <code class="docutils literal notranslate"><span class="pre">vb</span></code>.</p>
<p>Individuals in the VCFs are sampled (without replacement) for
inclusion in the output matrix. The size of the feature space
can therefore be vastly increased by having more individuals
in the VCFs than are needed for the feature dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vb</strong> (<a class="reference internal" href="#dinf.BagOfVcf" title="dinf.vcf.BagOfVcf"><em>dinf.vcf.BagOfVcf</em></a>) – The BagOfVcf object that describes the VCF/BCF files.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the genomic window to be sampled.</p></li>
<li><p><strong>max_missing_genotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Consider only sites with at most this many missing genotype calls.</p></li>
<li><p><strong>min_seg_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Sampled genotype matrix must have at least this many variable
sites (after filtering sites for missingness).</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Array with shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes,</span> <span class="pre">num_loci,</span> <span class="pre">c)</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="dinf.BinnedHaplotypeMatrix.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.BinnedHaplotypeMatrix.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Shape of the feature matrix.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.MultipleBinnedHaplotypeMatrices">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">MultipleBinnedHaplotypeMatrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_individuals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ploidy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_maf_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_phased</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/feature_extractor.html#MultipleBinnedHaplotypeMatrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.MultipleBinnedHaplotypeMatrices" title="Permalink to this definition">#</a></dt>
<dd><p>A factory for labelled collections of <a class="reference internal" href="#dinf.BinnedHaplotypeMatrix" title="dinf.BinnedHaplotypeMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">BinnedHaplotypeMatrix</span></code></a> objects.</p>
<p>One feature matrix is produced for each label. Labels correspond to
collections of individuals that will be treated as exchangeable
(e.g. populations).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_individuals</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the number of individuals
in the feature matrix.</p></li>
<li><p><strong>num_loci</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the number of feature loci to be
extracted from the sequence.</p></li>
<li><p><strong>ploidy</strong> (<em>Mapping</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>]</em>) – A dict that maps labels to the ploidy of the individuals.</p></li>
<li><p><strong>global_maf_thresh</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Minor allele frequency (MAF) threshold. Sites with MAF lower than
this value are ignored. MAF is calculated across all individuals.
If None, only invariant sites will be excluded.</p></li>
<li><p><strong>global_phased</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, the individuals’ haplotypes will each be included as
independent rows in each feature matrix and the shape of the
feature matrix for label <code class="docutils literal notranslate"><span class="pre">l</span></code> will be
<code class="docutils literal notranslate"><span class="pre">(ploidy[l]</span> <span class="pre">*</span> <span class="pre">num_individuals[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
If False, the allele counts for each individual will be summed
across their chromosome copies and the shape of the feature matrix
will be <code class="docutils literal notranslate"><span class="pre">(num_individuals[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.MultipleBinnedHaplotypeMatrices.from_ts">
<span class="sig-name descname"><span class="pre">from_ts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">individuals</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.MultipleBinnedHaplotypeMatrices.from_ts" title="Permalink to this definition">#</a></dt>
<dd><p>Create pseudo-genotype matrices from a tree sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><em>tskit.trees.TreeSequence</em></a>) – The tree sequence.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
<li><p><strong>individuals</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><em>Mapping</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)"><em>numpy.ndarray</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.10)"><em>Any</em></a><em>, </em><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v1.23)"><em>numpy.dtype</em></a><em>[</em><a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.integer" title="(in NumPy v1.23)"><em>numpy.integer</em></a><em>]</em><em>]</em><em>]</em>) – A mapping from label to an array of individuals.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary mapping a label <code class="docutils literal notranslate"><span class="pre">l</span></code> to a feature array.
Each array has shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
For an array <span class="math notranslate nohighlight">\(M\)</span>, the <span class="math notranslate nohighlight">\(M[i][j][0]\)</span>’th entry is the
count of minor alleles in the <span class="math notranslate nohighlight">\(j\)</span>’th bin of psdeudo-haplotype
<span class="math notranslate nohighlight">\(i\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.10)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.MultipleBinnedHaplotypeMatrices.from_vcf">
<span class="sig-name descname"><span class="pre">from_vcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vb</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_missing_genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_seg_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#dinf.MultipleBinnedHaplotypeMatrices.from_vcf" title="Permalink to this definition">#</a></dt>
<dd><p>Create pseudo-genotype matrices from a region of a VCF/BCF.</p>
<p>The genomic window is drawn uniformly at random from the sequences
defined in the given <a class="reference internal" href="#dinf.BagOfVcf" title="dinf.BagOfVcf"><code class="xref py py-class docutils literal notranslate"><span class="pre">BagOfVcf</span></code></a>.</p>
<p>Individuals in the VCFs are sampled (without replacement) for
inclusion in the output matrix. The size of the feature space
can therefore be vastly increased by having more individuals
in the VCFs than are needed for the feature dimensions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vb</strong> (<a class="reference internal" href="#dinf.BagOfVcf" title="dinf.vcf.BagOfVcf"><em>dinf.vcf.BagOfVcf</em></a>) – A collection of indexed VCF/BCF files.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the genomic window to be sampled.</p></li>
<li><p><strong>max_missing_genotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Consider only sites with fewer missing genotype calls than
this number.</p></li>
<li><p><strong>min_seg_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Sampled genotype matrix must have at least this many variable
sites (after filtering sites for missingness).</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary mapping a label <code class="docutils literal notranslate"><span class="pre">l</span></code> to a feature array.
Each array has shape <code class="docutils literal notranslate"><span class="pre">(num_pseudo_haplotypes[l],</span> <span class="pre">num_loci[l],</span> <span class="pre">c)</span></code>.
For an array <span class="math notranslate nohighlight">\(M\)</span>, the <span class="math notranslate nohighlight">\(M[i][j][0]\)</span>’th entry is the
count of minor alleles in the <span class="math notranslate nohighlight">\(j\)</span>’th bin of psdeudo-haplotype
<span class="math notranslate nohighlight">\(i\)</span>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Dict" title="(in Python v3.10)"><em>Dict</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="dinf.MultipleBinnedHaplotypeMatrices.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><span class="pre">int</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#dinf.MultipleBinnedHaplotypeMatrices.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Shape of the feature matrices.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.BagOfVcf">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">BagOfVcf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">files</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/vcf.html#BagOfVcf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.BagOfVcf" title="Permalink to this definition">#</a></dt>
<dd><p>A collection of indexed VCF or BCF files.</p>
<p>VCF data are sometimes contained in a single file, and sometimes split into
multiple files by chromosome. To remove the burden of dealing with both
of these common cases, this class maps a contig ID to a <a class="reference external" href="http://brentp.github.io/cyvcf2/docstrings.html#cyvcf2.cyvcf2.VCF" title="(in cyvcf2 v0.30.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">cyvcf2.VCF</span></code></a>
object for that contig. The class implements Python’s <a class="reference external" href="https://docs.python.org/3/glossary.html#term-mapping" title="(in Python v3.10)"><span>mapping</span></a>
protocol, plus methods for sampling regions of the genome at random.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">dinf</span>

<span class="n">vcfs</span> <span class="o">=</span> <span class="n">dinf</span><span class="o">.</span><span class="n">BagOfVcf</span><span class="p">([</span><span class="s2">&quot;chr1.bcf&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2.bcf&quot;</span><span class="p">,</span> <span class="s2">&quot;chr3.bcf&quot;</span><span class="p">])</span>

<span class="c1"># Iterate over contig IDs.</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">vcfs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span>
<span class="k">assert</span> <span class="nb">list</span><span class="p">(</span><span class="n">vcfs</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="s2">&quot;chr3&quot;</span><span class="p">]</span>

<span class="c1"># Lookup a cyvcf2.VCF object by contig ID.</span>
<span class="n">chr1</span> <span class="o">=</span> <span class="n">vcfs</span><span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">]</span>
<span class="k">assert</span> <span class="n">chr1</span><span class="o">.</span><span class="n">fname</span> <span class="o">==</span> <span class="s2">&quot;chr1.bcf&quot;</span>
<span class="n">first_variant</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">chr1</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">first_variant</span><span class="o">.</span><span class="n">CHROM</span> <span class="o">==</span> <span class="s2">&quot;chr1&quot;</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>files</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a><em>]</em>) – An iterable of filenames. Each file must be an indexed VCF or BCF,
and must contain the FORMAT/GT field.</p></li>
<li><p><strong>contig_lengths</strong> (<a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.10)"><em>collections.abc.Mapping</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>] </em><em>| </em><em>None</em>) – A dict mapping a contig name to contig length. Only the contigs in
this dict will be used.</p></li>
<li><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.10)"><em>collections.abc.Mapping</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><em>List</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>]</em><em>] </em><em>| </em><em>None</em>) – A dictionary that maps a label to a list of individual names,
where the individual names correspond to the VCF columns
for which genotypes will be sampled.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.BagOfVcf.lengths">
<span class="sig-name descname"><span class="pre">lengths</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">np.ndarray</span></em><a class="headerlink" href="#dinf.BagOfVcf.lengths" title="Permalink to this definition">#</a></dt>
<dd><p>Lengths of the contigs in the bag.</p>
<p>The order matches the contig order obtained when iterating.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="dinf.BagOfVcf.samples">
<span class="sig-name descname"><span class="pre">samples</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.10)"><span class="pre">collections.abc.Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.10)"><span class="pre">None</span></a></em><a class="headerlink" href="#dinf.BagOfVcf.samples" title="Permalink to this definition">#</a></dt>
<dd><p>A dictionary that maps a label to a list of individual names.</p>
<p>The individual names correspond to the VCF columns for which genotypes
will be sampled. This is a bookkeeping device that records which genotypes
belong to which label (e.g. which population). If None, it is assumed
that all individuals in the VCF will be treated as exchangeable.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.BagOfVcf.sample_regions">
<span class="sig-name descname"><span class="pre">sample_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/vcf.html#BagOfVcf.sample_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.BagOfVcf.sample_regions" title="Permalink to this definition">#</a></dt>
<dd><p>Sample a list of (chrom, start, end) triplets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of genomic windows to sample.</p></li>
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the sequence to sample.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – The numpy random number generator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of 3-tuples: (chrom, start, end).
The start and end coordinates are 1-based and inclusive, to match
the usual convention for ‘chrom:start-end’, e.g. in bcftools.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.10)"><em>List</em></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)">int</a>]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.BagOfVcf.sample_genotype_matrix">
<span class="sig-name descname"><span class="pre">sample_genotype_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_seg_sites</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_missing_genotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_phased</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rng</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/vcf.html#BagOfVcf.sample_genotype_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.BagOfVcf.sample_genotype_matrix" title="Permalink to this definition">#</a></dt>
<dd><p>Sample a genotype matrix uniformly at random from the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Length of the sequence to sample.</p></li>
<li><p><strong>max_missing_genotypes</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Consider only sites with at most this many missing genotype calls.</p></li>
<li><p><strong>min_seg_sites</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Sampled genotype matrix must have at least this many variable
sites (after filtering sites for missingness).</p></li>
<li><p><strong>require_phased</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, raise an error if genotypes are not phased.</p></li>
<li><p><strong>rng</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator" title="(in NumPy v1.23)"><em>numpy.random.Generator</em></a>) – Numpy random number generator.</p></li>
<li><p><strong>retries</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Maximum number of attempts allowed to find a genomic window with
the required number of segregating sites.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>A 2-tuple of (G, positions) where</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">G</span></code> is a (num_sites, num_individuals, ploidy) genotype matrix,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">positions</span></code> are the site coordinates, as an offset from the
the start of the genomic window.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.23)">numpy.ndarray</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="sec-api-train">
<span id="id4"></span><h3>Training<a class="headerlink" href="#sec-api-train" title="Permalink to this headline">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="dinf.train">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">train</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dinf_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallelism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#train"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.train" title="Permalink to this definition">#</a></dt>
<dd><p>Train a discriminator network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dinf_model</strong> (<a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><em>DinfModel</em></a>) – DinfModel object that describes the model components.</p></li>
<li><p><strong>training_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the dataset used to train the discriminator.</p></li>
<li><p><strong>test_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the test dataset used to evaluate the discriminator after
each training epoch.</p></li>
<li><p><strong>epochs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of full passes over the training dataset when training
the discriminator.</p></li>
<li><p><strong>parallelism</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of processes to use for parallelising calls to the
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.generator_func()</span></code></a> and
<a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.target_func()</span></code></a>.</p></li>
<li><p><strong>seed</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Seed for the random number generator.</p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The trained discriminator.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#dinf.Discriminator" title="dinf.Discriminator">Discriminator</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="sec-api-inference">
<span id="id5"></span><h3>Inference<a class="headerlink" href="#sec-api-inference" title="Permalink to this headline">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="dinf.predict">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">discriminator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dinf_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallelism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.predict" title="Permalink to this definition">#</a></dt>
<dd><p>Sample features and make predictions using the discriminator.</p>
<p>Features are sampled from the generator by default.
To instead sample features from the target dataset, use the
<code class="docutils literal notranslate"><span class="pre">sample_target</span></code> option.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dinf_model</strong> (<a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><em>DinfModel</em></a>) – DinfModel object that describes the model components.</p></li>
<li><p><strong>replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of features to extract.</p></li>
<li><p><strong>sample_target</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, sample features from the target dataset.
If False (the default), features are sampled from the generator.</p></li>
<li><p><strong>parallelism</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of processes to use for parallelising calls to the
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.generator_func()</span></code></a> and
<a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.target_func()</span></code></a>.</p></li>
<li><p><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em> | </em><em>None</em>) – Seed for the random number generator.</p></li>
<li><p><strong>discriminator</strong> (<a class="reference internal" href="#dinf.Discriminator" title="dinf.Discriminator"><em>Discriminator</em></a>) – </p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A 2-tuple of (thetas, probs), where <code class="docutils literal notranslate"><span class="pre">thetas</span></code> are the drawn parameters
(or <code class="docutils literal notranslate"><span class="pre">None</span></code> if <code class="docutils literal notranslate"><span class="pre">sample_target</span></code> is True) and <code class="docutils literal notranslate"><span class="pre">probs</span></code> are the
discriminator predictions.
thetas[j][k] is the j’th draw for the k’th parameter, and
probs[j] is the discriminator prediction for the j’th draw.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Tuple[np.ndarray | None, np.ndarray]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.abc_gan">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">abc_gan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dinf_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proposal_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallelism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#abc_gan"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.abc_gan" title="Permalink to this definition">#</a></dt>
<dd><p>Adversarial Abstract Bayesian Computation.</p>
<p>Conceptually, the GAN takes the following steps for iteration j:</p>
<blockquote>
<div><ul class="simple">
<li><p>sample training and proposal datasets from the prior[j] distribution,</p></li>
<li><p>train the discriminator,</p></li>
<li><p>make predictions with the discriminator on the proposal dataset,</p></li>
<li><p>construct a posterior[j] sample from the proposal dataset,</p></li>
<li><p>set prior[j+1] = posterior[j].</p></li>
</ul>
</div></blockquote>
<p>In the first iteration, the parameter values given to the generator
to produce the train/proposal datasets are drawn from the parameters’
prior distribution. In subsequent iterations, the parameter values
are drawn from a posterior ABC sample. The posterior is obtained by
rejection sampling the proposal distribution and weighting the posterior
by the discriminator predictions, followed by gaussian smoothing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dinf_model</strong> (<a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><em>DinfModel</em></a>) – DinfModel object that describes the dinf model.</p></li>
<li><p><strong>iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of GAN iterations.</p></li>
<li><p><strong>training_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the dataset used to train the discriminator.
This dataset is constructed once each GAN iteration.</p></li>
<li><p><strong>test_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the dataset used to evaluate the discriminator after
each training epoch. This dataset is constructed once before the
GAN iterates, and is reused in each iteration.</p></li>
<li><p><strong>proposal_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of ABC proposals in each iteration.</p></li>
<li><p><strong>epochs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of full passes over the training dataset when training
the discriminator.</p></li>
<li><p><strong>top_n</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em> | </em><em>None</em>) – If not None, do ABC rejection sampling in each iteraction by taking
the <code class="docutils literal notranslate"><span class="pre">top_n</span></code> best samples, ranked by discriminator prediction.
Samples are taken from the test replicates, so <code class="docutils literal notranslate"><span class="pre">top_n</span></code> must be
smaller than <code class="docutils literal notranslate"><span class="pre">test_replicates</span></code>.</p></li>
<li><p><strong>output_folder</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – Folder to output results. If not specified, the current
directory will be used.</p></li>
<li><p><strong>parallelism</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of processes to use for parallelising calls to the
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.generator_func()</span></code></a> and
<a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.target_func()</span></code></a>.</p></li>
<li><p><strong>seed</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Seed for the random number generator.</p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.mcmc_gan">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">mcmc_gan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dinf_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">walkers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">steps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Dx_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallelism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#mcmc_gan"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.mcmc_gan" title="Permalink to this definition">#</a></dt>
<dd><blockquote>
<div><p>Run the MCMC GAN.</p>
</div></blockquote>
<p>Conceptually, the GAN takes the following steps for iteration j:</p>
<blockquote>
<div><ul class="simple">
<li><p>sample training dataset from the prior[j] distribution,</p></li>
<li><p>train the discriminator,</p></li>
<li><p>run the MCMC,</p></li>
<li><p>obtain posterior[j] as weighted KDE of MCMC sample,</p></li>
<li><p>set prior[j+1] = posterior[j].</p></li>
</ul>
</div></blockquote>
<p>In the first iteration, the parameter values given to the generator
to produce the training dataset are drawn from the parameters’ prior
distribution. In subsequent iterations, the parameter values are drawn
from a weighted gaussian KDE of the previous iteration’s MCMC chains.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dinf_model</strong> (<a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><em>DinfModel</em></a>) – DinfModel object that describes the model components.</p></li>
<li><p><strong>iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of GAN iterations.</p></li>
<li><p><strong>training_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the dataset used to train the discriminator.
This dataset is constructed once each GAN iteration.</p></li>
<li><p><strong>test_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the test dataset used to evaluate the discriminator after
each training epoch.</p></li>
<li><p><strong>epochs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of full passes over the training dataset when training
the discriminator.</p></li>
<li><p><strong>walkers</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of independent MCMC chains.</p></li>
<li><p><strong>steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The chain length for each MCMC walker.</p></li>
<li><p><strong>Dx_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of generator replicates for approximating E[D(x)|θ].</p></li>
<li><p><strong>output_folder</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – Folder to output results. If not specified, the current
directory will be used.</p></li>
<li><p><strong>parallelism</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of processes to use for parallelising calls to the
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.generator_func()</span></code></a> and
<a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.target_func()</span></code></a>.</p></li>
<li><p><strong>seed</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Seed for the random number generator.</p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.pg_gan">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">pg_gan</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dinf_model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">training_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Dx_replicates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_proposals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proposal_stddev</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.06666666666666667</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pretraining_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dinf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proposals_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_pretraining_iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallelism</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#pg_gan"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.pg_gan" title="Permalink to this definition">#</a></dt>
<dd><p>PG-GAN style simulated annealing.</p>
<p>Wang et al. 2021, <a class="reference external" href="https://doi.org/10.1111/1755-0998.13386">https://doi.org/10.1111/1755-0998.13386</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dinf_model</strong> (<a class="reference internal" href="#dinf.DinfModel" title="dinf.DinfModel"><em>DinfModel</em></a>) – DinfModel object that describes the dinf model.</p></li>
<li><p><strong>iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of GAN iterations.</p></li>
<li><p><strong>training_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the dataset used to train the discriminator.
This dataset is constructed once each GAN iteration.</p></li>
<li><p><strong>test_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of the test dataset used to evaluate the discriminator after
each training epoch.</p></li>
<li><p><strong>epochs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of full passes over the training dataset when training
the discriminator.</p></li>
<li><p><strong>Dx_replicates</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of generator replicates for approximating E[D(x)|θ].</p></li>
<li><p><strong>num_proposals</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number proposals per parameter in each iteration.</p></li>
<li><p><strong>proposal_stddev</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a>) – The standard deviation for the proposal distribution is:
temperature * proposal_stddev * (param.high - param.low)
The default is 1 / 15, as used by PG-GAN.</p></li>
<li><p><strong>pretraining_method</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – <p>A string indicating which method should be used to pretrain the
discriminator and choose a starting point in parameter space.</p>
<blockquote>
<div><ul>
<li><p>”dinf” (<em>default</em>):
Train the discriminator on data sampled from the prior, until it can
distinguish randomly chosen points from the target dataset with
accuracy 0.9, or until <code class="docutils literal notranslate"><span class="pre">max_pretraining_iterations</span></code> is exhausted.
In practice, 10,000–20,000 training instances are sufficient.</p>
<p>After pretraining, we choose the starting point to be the point
with the highest log probability from a fresh set of candidates
drawn from the prior.</p>
</li>
<li><p>”pg-gan”:
PG-GAN starts at a point in the parameter space that is favourable
to the discriminator, and it only trains the discriminator enough
to identify such a point.</p>
<p>In each iteration we train the discriminator on a single point
sampled from the prior. If the discriminator can distinguish between
that point and the target dataset with accuracy 0.9, this point is
used as the starting point.
Otherwise, a new point is chosen by sampling from the prior.
If <code class="docutils literal notranslate"><span class="pre">max_pretraining_iterations</span></code> is exhausted without reaching an
accuracy of 0.9, the point at which the discriminator had the
highest accuracy is used as the starting point.</p>
</li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>proposals_method</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – <p>A string indicating which method should be used to produce the proposal
distribution. In each case, <code class="docutils literal notranslate"><span class="pre">num_proposals</span></code> × P proposals are
produced, where P is the number of parameters (to match original
PG-GAN behaviour).</p>
<blockquote>
<div><ul>
<li><p>”rr” (<em>default</em>):
Round-robin proposals. Proposals are generated for one parameter
only, keeping the other parameters fixed.  A different parameter is
chosen in each iteration, cycling through all parameters in order.
Proposals are drawn from a normal distribution with standard deviation:
temperature × <code class="docutils literal notranslate"><span class="pre">proposal_stddev</span></code> × (param.high - param.low).</p></li>
<li><p>”mvn”:
Multivariate-normal proposals.
Proposals are drawn from a multivariate normal distribution with
covariance matrix defined to have diagonals:
(temperature × <code class="docutils literal notranslate"><span class="pre">proposal_stddev</span></code> × (param.high - param.low))**2 / P,
where P is the number of parameters and the high/low bounds vary by
the corresponding parameter. Off-diagonals of the covariance matrix
are set to zero.</p></li>
<li><p>”pg-gan”:
Proposals like PG-GAN.
<code class="docutils literal notranslate"><span class="pre">num_proposals</span></code> proposals are generated for each parameter,
keeping the other parameters fixed, thus producing
<code class="docutils literal notranslate"><span class="pre">num_proposals</span></code> × P total proposals (for P parameters).
Proposals are drawn from a normal distribution with standard deviation:
temperature × <code class="docutils literal notranslate"><span class="pre">proposal_stddev</span></code> × (param.high - param.low).</p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>max_pretraining_iterations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The maximum number of pretraining iterations.</p></li>
<li><p><strong>output_folder</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – Folder to output results. If not specified, the current
directory will be used.</p></li>
<li><p><strong>parallelism</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Number of processes to use for parallelising calls to the
<a class="reference internal" href="#dinf.DinfModel.generator_func" title="dinf.DinfModel.generator_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.generator_func()</span></code></a> and
<a class="reference internal" href="#dinf.DinfModel.target_func" title="dinf.DinfModel.target_func"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DinfModel.target_func()</span></code></a>.</p></li>
<li><p><strong>seed</strong> (<em>None</em><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Seed for the random number generator.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="sec-api-classification">
<span id="id6"></span><h3>Classification<a class="headerlink" href="#sec-api-classification" title="Permalink to this headline">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="dinf.Discriminator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">Discriminator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trained=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metrics=&lt;factory&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format_version=4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_inited=False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator" title="Permalink to this definition">#</a></dt>
<dd><p>Wrapper for the discriminator neural network that classifies genotype matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>network</strong> (<em>nn.Module</em><em> | </em><em>None</em>) – </p></li>
<li><p><strong>state</strong> (<em>TrainState</em><em> | </em><em>None</em>) – </p></li>
<li><p><strong>trained</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – </p></li>
<li><p><strong>metrics</strong> (<a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping" title="(in Python v3.10)"><em>collections.abc.Mapping</em></a>) – </p></li>
<li><p><strong>format_version</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – </p></li>
<li><p><strong>_inited</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.Discriminator.from_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator.from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator.from_file" title="Permalink to this definition">#</a></dt>
<dd><p>Load neural network weights from the given file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – The filename of the saved model.</p></li>
<li><p><strong>network</strong> – The flax neural network.</p></li>
<li><p><strong>state</strong> – The flax train state.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new Discriminator object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Discriminator.to_file">
<span class="sig-name descname"><span class="pre">to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator.to_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator.to_file" title="Permalink to this definition">#</a></dt>
<dd><p>Save neural network to the given file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – The path where the model will be saved.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Discriminator.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ss</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_regularisation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator.fit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator.fit" title="Permalink to this definition">#</a></dt>
<dd><p>Fit discriminator to training data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>train_x</strong> – Training data.</p></li>
<li><p><strong>train_y</strong> – Labels for training data (zeros and ones).</p></li>
<li><p><strong>val_x</strong> – Validation data.</p></li>
<li><p><strong>val_y</strong> – Labels for validation data (zeros and ones).</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of minibatch for gradient update step.</p></li>
<li><p><strong>epochs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The number of full passes over the training data.</p></li>
<li><p><strong>ss</strong> (<a class="reference external" href="https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.SeedSequence.html#numpy.random.SeedSequence" title="(in NumPy v1.23)"><em>numpy.random.SeedSequence</em></a>) – Numpy random number seed sequence.</p></li>
<li><p><strong>reset_metrics</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If true, remove loss/accuracy metrics from previous calls to
fit() (if any). If false, loss/accuracy metrics will be appended
to the existing metrics.</p></li>
<li><p><strong>entropy_regularisation</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – </p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Discriminator.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1024</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator.predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator.predict" title="Permalink to this definition">#</a></dt>
<dd><p>Make predictions about data using a pre-fitted neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> – The data instances about which to make predictions.</p></li>
<li><p><strong>batch_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Size of data batches for prediction.</p></li>
<li><p><strong>callbacks</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A vector of predictions, one for each input instance.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Discriminator.summary">
<span class="sig-name descname"><span class="pre">summary</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Discriminator.summary"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Discriminator.summary" title="Permalink to this definition">#</a></dt>
<dd><p>Return a summary of the neural network.</p>
</dd></dl>

</dd></dl>

</section>
<section id="sec-api-discriminator-networks">
<span id="id7"></span><h3>Discriminator networks<a class="headerlink" href="#sec-api-discriminator-networks" title="Permalink to this headline">#</a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="dinf.ExchangeableCNN">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">ExchangeableCNN</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sizes1=(32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symmetric=Symmetric(</span>&#160;&#160;&#160;&#160; <span class="pre">#</span> <span class="pre">attributes</span>&#160;&#160;&#160;&#160; <span class="pre">func</span> <span class="pre">=</span> <span class="pre">'max'</span>&#160;&#160;&#160;&#160; <span class="pre">k</span> <span class="pre">=</span> <span class="pre">None</span> <span class="pre">)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizes2=(64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent=&lt;flax.linen.module._Sentinel</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name=None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#ExchangeableCNN"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.ExchangeableCNN" title="Permalink to this definition">#</a></dt>
<dd><p>An exchangeable CNN for the discriminator.</p>
<p>Each feature matrix in the input has shape
(batch_size, num_haplotypes, num_loci, channels).
Two 1-d convolution layers are applied along haplotypes and
feature size is reduced in each convolution using a stride of 2.
Then the max function is used to collapse
features across the num_haplotypes dimension. A third 1-d convolution
follows (with stride of 1), and the max function is used to
collapse features across the num_loci dimension.
Elu activation is used throughout, except for the final output which has
no activation.</p>
<p>The number of trainable parameters in the network is dependent on the
number of feature matrices in the input, but independent of the size of
the feature matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sizes1</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><em>...</em><em>]</em>) – A tuple of feature sizes for the convolution layers before the
symmetric function.</p></li>
<li><p><strong>symmetric</strong> (<a class="reference external" href="https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module" title="(in Flax)"><em>flax.linen.module.Module</em></a>) – The symmetric network layer.</p></li>
<li><p><strong>sizes2</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><em>...</em><em>]</em>) – A tuple of feature sizes for the convolution layers afer the
symmetric function.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><em>Optional</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><em>Union</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><a class="reference external" href="https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module" title="(in Flax)"><em>flax.linen.module.Module</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.core.scope.Scope</em><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.linen.module._Sentinel</em><em>]</em><em>]</em><em>]</em>) – </p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.ExchangeablePGGAN">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">ExchangeablePGGAN</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sizes1=(32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">64)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symmetric=Symmetric(</span>&#160;&#160;&#160;&#160; <span class="pre">#</span> <span class="pre">attributes</span>&#160;&#160;&#160;&#160; <span class="pre">func</span> <span class="pre">=</span> <span class="pre">'sum'</span>&#160;&#160;&#160;&#160; <span class="pre">k</span> <span class="pre">=</span> <span class="pre">None</span> <span class="pre">)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizes2=(128</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">128)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent=&lt;flax.linen.module._Sentinel</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name=None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#ExchangeablePGGAN"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.ExchangeablePGGAN" title="Permalink to this definition">#</a></dt>
<dd><p>An exchangeable CNN for the discriminator using the PG-GAN architecture.</p>
<p>This is a faithful translation of the network implemented in PG-GAN [1],
with the addition of batch normalisation.
Each haplotype is treated as exchangeable with any other
haplotype within the same feature matrix (i.e. exchangeability
applies within labelled groups).</p>
<p>Each feature matrix in the input has shape
(batch_size, num_haplotypes, num_loci, channels).
Two 1-d convolution layers are applied along haplotypes, and feature
size is reduced following each convolution using 1-d max pooling with
size and stride of 2. The sum function is then used to collapse
features across the num_haplotypes dimension. Two fully connected layers
with dropout follow.
Relu activation is used throughout, except for the final output which has
no activation.</p>
<p>The number of trainable parameters in the network is dependent on the
number of feature matrices in the input, and dependent on the num_loci
dimension of the feature matrices.</p>
<div class="line-block">
<div class="line">[1] Wang et al. 2021, <a class="reference external" href="https://doi.org/10.1111/1755-0998.13386">https://doi.org/10.1111/1755-0998.13386</a></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sizes1</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><em>...</em><em>]</em>) – A tuple of feature sizes for the convolution layers before the
symmetric function. Convolution weights are tied between convolution
layers of the distinct labelled feature matrices.</p></li>
<li><p><strong>symmetric</strong> (<a class="reference external" href="https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module" title="(in Flax)"><em>flax.linen.module.Module</em></a>) – The symmetric network layer.</p></li>
<li><p><strong>sizes2</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><em>Tuple</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em>, </em><em>...</em><em>]</em>) – A tuple of feature sizes for the dense (fully connected) layers
afer the symmetric function.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><em>Optional</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><em>Union</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><a class="reference external" href="https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module" title="(in Flax)"><em>flax.linen.module.Module</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.core.scope.Scope</em><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.linen.module._Sentinel</em><em>]</em><em>]</em><em>]</em>) – </p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.Symmetric">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">Symmetric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">func</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent=&lt;flax.linen.module._Sentinel</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name=None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/discriminator.html#Symmetric"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Symmetric" title="Permalink to this definition">#</a></dt>
<dd><p>Network layer that summarises over a given axis in a way that is invariant
to permutations of the input along that axis.</p>
<div class="line-block">
<div class="line">Chan et al. 2018, <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7687905/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7687905/</a></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – <p>The permutation-invariant function to apply.
Must be one of:</p>
<blockquote>
<div><ul>
<li><p>”max”: maximum value.</p></li>
<li><p>”mean”: mean value.</p></li>
<li><p>”sum”: sum of the values.</p></li>
<li><p>”var”: variance of the values.</p></li>
<li><p>”moments”: first <code class="docutils literal notranslate"><span class="pre">k</span></code> moments, <span class="math notranslate nohighlight">\(E[x], E[x^2], ..., E[x^k]\)</span>.</p></li>
<li><p>”central-moments”: first <code class="docutils literal notranslate"><span class="pre">k</span></code> central moments,
<span class="math notranslate nohighlight">\(E[x], E[(x - E[x])^2], ..., E[(x - E[x])^k]\)</span>.</p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>k</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em> | </em><em>None</em>) – The number of moments to take when <code class="docutils literal notranslate"><span class="pre">func</span></code> is “moments” or “central-moments”.</p></li>
<li><p><strong>parent</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><em>Optional</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><em>Union</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><a class="reference external" href="https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module" title="(in Flax)"><em>flax.linen.module.Module</em></a><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.core.scope.Scope</em><em>]</em><em>, </em><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Type" title="(in Python v3.10)"><em>Type</em></a><em>[</em><em>flax.linen.module._Sentinel</em><em>]</em><em>]</em><em>]</em>) – </p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="sec-api-misc">
<span id="id8"></span><h3>Miscellaneous<a class="headerlink" href="#sec-api-misc" title="Permalink to this headline">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="dinf.save_results">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">save_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thetas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">probs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameters</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#save_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.save_results" title="Permalink to this definition">#</a></dt>
<dd><p>Save discriminator predictions and parameter values to a .npz file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – The output file.</p></li>
<li><p><strong>thetas</strong> (<em>np.ndarray</em><em> | </em><em>None</em>) – Parameter values.</p></li>
<li><p><strong>probs</strong> (<em>np.ndarray</em>) – Discriminator predictions corresponding to the <code class="docutils literal notranslate"><span class="pre">thetas</span></code>.</p></li>
<li><p><strong>parameters</strong> (<a class="reference internal" href="#dinf.Parameters" title="dinf.Parameters"><em>Parameters</em></a>) – Sequence of parameter names.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.load_results">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">load_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/dinf.html#load_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.load_results" title="Permalink to this definition">#</a></dt>
<dd><p>Load descriminator predictions and parameter values from a .npz file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – The input file.</p></li>
<li><p><strong>parameters</strong> (<a class="reference internal" href="#dinf.Parameters" title="dinf.Parameters"><em>Parameters</em></a><em> | </em><em>None</em>) – Sequence of parameter names against which the file’s
arrays will be checked. A ValueError exception is raised
if the names are not the same (and in the same order).
If None, the parameter names are not checked.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A numpy structured array, where the first column is the probabilities
(named <code class="docutils literal notranslate"><span class="pre">_Pr</span></code>), and the subsequence columns are the
parameter values (if any).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.ts_individuals">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">ts_individuals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ts</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/misc.html#ts_individuals"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.ts_individuals" title="Permalink to this definition">#</a></dt>
<dd><p>Get the individuals corresponding to the tree sequence’s samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ts</strong> (<a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><em>tskit.TreeSequence</em></a>) – The tree sequence.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a><em> | </em><em>None</em>) – Only return individuals from this population. The population may be
a string identifier (which will be matched against the ‘name’
metadata field in the population table of the tree sequence),
or an integer population id.
If not specified, all sampled individuals will be returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of individual IDs (indices into the individuals table).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>npt.NDArray[np.integer]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.get_samples_from_1kgp_metadata">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">get_samples_from_1kgp_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/vcf.html#get_samples_from_1kgp_metadata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.get_samples_from_1kgp_metadata" title="Permalink to this definition">#</a></dt>
<dd><p>Get sample IDs for 1000 Genomes Project populations.
Related individuals are removed based on the FatherID and MotherID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – Path to the file “20130606_g1k_3202_samples_ped_population.txt”
This file can be downloaded from
<a class="reference external" href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/">http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/</a></p></li>
<li><p><strong>populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.10)"><em>list</em></a>) – List of populations to extract.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict mapping population names to a list of sample IDs.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.get_contig_lengths">
<span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">get_contig_lengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_contigs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/vcf.html#get_contig_lengths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.get_contig_lengths" title="Permalink to this definition">#</a></dt>
<dd><p>Load contig lengths from a whitespace-separated file like an fai (fasta index).</p>
<p>The file must have (at least) two columns, the first specifies the
contig ID, and the second is the contig length. Additional columns
are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a>) – The path to the file.</p></li>
<li><p><strong>keep_contigs</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>] </em><em>| </em><em>None</em>) – Use only the specified contigs.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)">dict</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="dinf.Store">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">dinf.</span></span><span class="sig-name descname"><span class="pre">Store</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/store.html#Store"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Store" title="Permalink to this definition">#</a></dt>
<dd><p>A sequence of folders with names “0”, “1”, …</p>
<p>This class implements the <a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.abc.Sequence</span></code></a> protocol.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><p><strong>base</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – Base directory containing the sequence.</p>
</dd>
<dt class="field-even">Parameters</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>base</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.10)"><em>pathlib.Path</em></a>) – The base directory containing the sequence.</p></li>
<li><p><strong>create</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, create the base directory if it doesn’t exist.
If False (<em>default</em>), raise an error if the base directory
doesn’t exist.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="dinf.Store.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/store.html#Store.__len__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Store.__len__" title="Permalink to this definition">#</a></dt>
<dd><p>Length of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Store.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/store.html#Store.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Store.__getitem__" title="Permalink to this definition">#</a></dt>
<dd><p>Get the path with the given index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>index</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – The 0-based index into the sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Store.increment">
<span class="sig-name descname"><span class="pre">increment</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/store.html#Store.increment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Store.increment" title="Permalink to this definition">#</a></dt>
<dd><p>Add a new folder to the end of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="dinf.Store.assert_complete">
<span class="sig-name descname"><span class="pre">assert_complete</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">expected_files</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/store.html#Store.assert_complete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.Store.assert_complete" title="Permalink to this definition">#</a></dt>
<dd><p>Raise a RuntimeError if any expected files are missing from a subfolder.</p>
</dd></dl>

</dd></dl>

</section>
<section id="sec-api-plotting">
<span id="id9"></span><h3>Plotting<a class="headerlink" href="#sec-api-plotting" title="Permalink to this headline">#</a></h3>
<span class="target" id="module-dinf.plot"></span><dl class="py function">
<dt class="sig sig-object py" id="dinf.plot.feature">
<span class="sig-prename descclassname"><span class="pre">dinf.plot.</span></span><span class="sig-name descname"><span class="pre">feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/plot.html#feature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.plot.feature" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a feature matrix as a heatmap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mat</strong> (<em>np.ndarray</em>) – The feature matrix.</p></li>
<li><p><strong>ax</strong> (<a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)"><em>matplotlib.axes.Axes</em></a>) – Matplotlib axes onto which the plot will be drawn.</p></li>
<li><p><strong>channel</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><em>int</em></a>) – Which channel of the feature matrix to draw.
I.e., the index into the last dimension of <code class="docutils literal notranslate"><span class="pre">mat</span></code>.</p></li>
<li><p><strong>cmap</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><em>matplotlib.colors.Colourmap</em><em> | </em><em>None</em>) – Matplotlib colour map.</p></li>
<li><p><strong>cb</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, add a colour bar that indicates how colours relate to the
values in the feature matrix.</p></li>
<li><p><strong>cb_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><em>None</em>) – Label on the colour bar.</p></li>
<li><p><strong>vmax</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.plot.features">
<span class="sig-prename descclassname"><span class="pre">dinf.plot.</span></span><span class="sig-name descname"><span class="pre">features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mats</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subplots_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/plot.html#features"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.plot.features" title="Permalink to this definition">#</a></dt>
<dd><p>Plot multiple feature matrices as heatmaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mats</strong> (<em>Dict</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><em>np.ndarray</em><em>]</em>) – The multiple feature matrices.</p></li>
<li><p><strong>subplots_kw</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – Additional keyword arguments that will be passed to
<a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplots.html#matplotlib.pyplot.subplots" title="(in Matplotlib v3.6.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.subplots()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>[<a class="reference external" href="https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure" title="(in Matplotlib v3.6.0)">matplotlib.figure.Figure</a>, <a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)">matplotlib.axes.Axes</a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A (figure, axes) tuple.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.plot.metrics">
<span class="sig-prename descclassname"><span class="pre">dinf.plot.</span></span><span class="sig-name descname"><span class="pre">metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metrics_collection</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subplot_mosaic_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/plot.html#metrics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.plot.metrics" title="Permalink to this definition">#</a></dt>
<dd><p>Plot training metrics from a discriminator neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>metrics_collection</strong> (<em>Dict</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><em>Dict</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em>, </em><em>Any</em><em>]</em><em>]</em>) – A dictionary mapping labels to metrics dictionaries.
The labels will be used in the legend to identify each
discriminator.</p></li>
<li><p><strong>legend_title</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><em>str</em></a><em> | </em><em>None</em>) – A title for the legend.</p></li>
<li><p><strong>subplot_mosaic_kw</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – Additional keyword arguments that will be passed to
<a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplot_mosaic.html#matplotlib.pyplot.subplot_mosaic" title="(in Matplotlib v3.6.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.subplot_mosaic()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>[<a class="reference external" href="https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure" title="(in Matplotlib v3.6.0)">matplotlib.figure.Figure</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)">str</a>, <a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)">matplotlib.axes.Axes</a>]]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A (figure, axes_dict) tuple, where axes_dict is a dictionary mapping
metrics to the <a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">matplotlib.axes.Axes</span></code></a> objects.
(As obtained from <a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplot_mosaic.html#matplotlib.pyplot.subplot_mosaic" title="(in Matplotlib v3.6.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.subplot_mosaic()</span></code></a>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.plot.hist2d">
<span class="sig-prename descclassname"><span class="pre">dinf.plot.</span></span><span class="sig-name descname"><span class="pre">hist2d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hist2d_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/plot.html#hist2d"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.plot.hist2d" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a 2d histogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>np.ndarray</em>) – Values for the horizontal axis.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Values for the vertical axis.</p></li>
<li><p><strong>x_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Name of the parameter on the horizontal axis.</p></li>
<li><p><strong>y_label</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – Name of the parameter on the vertical axis.</p></li>
<li><p><strong>x_truth</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – True value of the parameter on the horizontal axis.</p></li>
<li><p><strong>y_truth</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – True value of the parameter on the vertical axis.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axis</em><em> | </em><em>None</em>) – Axes onto which the histogram will be drawn.
If None, an axes will be created with <a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplots.html#matplotlib.pyplot.subplots" title="(in Matplotlib v3.6.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.subplots()</span></code></a>.</p></li>
<li><p><strong>hist2d_kw</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – Further parameters passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.axes.Axes.hist2d()</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>[<a class="reference external" href="https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure" title="(in Matplotlib v3.6.0)">matplotlib.figure.Figure</a>, <a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)">matplotlib.axes.Axes</a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A (figure, axes) tuple.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="dinf.plot.hist">
<span class="sig-prename descclassname"><span class="pre">dinf.plot.</span></span><span class="sig-name descname"><span class="pre">hist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ci</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hist_kw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dinf/plot.html#hist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#dinf.plot.hist" title="Permalink to this definition">#</a></dt>
<dd><p>Plot a histogram.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> (<em>np.ndarray</em>) – Array of values for the histogram.</p></li>
<li><p><strong>ax</strong> (<a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)"><em>matplotlib.axes.Axes</em></a><em> | </em><em>None</em>) – Axes onto which the histogram will be drawn.
If None, an axes will be created with <a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplots.html#matplotlib.pyplot.subplots" title="(in Matplotlib v3.6.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.pyplot.subplots()</span></code></a>.</p></li>
<li><p><strong>ci</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><em>bool</em></a>) – If True, draw a 95% credible interval at the bottom.</p></li>
<li><p><strong>truth</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><em>float</em></a><em> | </em><em>None</em>) – If not None, draw a red vertical line at this location.</p></li>
<li><p><strong>hist_kw</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.10)"><em>dict</em></a><em> | </em><em>None</em>) – Further parameters passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">matplotlib.axes.Axes.hist()</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.10)">tuple</a>[<a class="reference external" href="https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure" title="(in Matplotlib v3.6.0)">matplotlib.figure.Figure</a>, <a class="reference external" href="https://matplotlib.org/stable/api/axes_api.html#matplotlib.axes.Axes" title="(in Matplotlib v3.6.0)">matplotlib.axes.Axes</a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A (figure, axes) tuple.</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="guide/pg-gan.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">PG-GAN (simulated annealing)</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="cli.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">CLI reference</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Graham Gower<br/>
  
      &copy; Copyright 2021-2022.<br/>
    <div class="extra_footer">
      dinf 0.3.1.dev58+g0937a89
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>