
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dinf.feature_extractor &#8212; Dinf</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Dinf</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../introduction.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../tutorial/01.html">
   1. Create a Dinf model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../api.html">
   API reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cli.html">
   CLI reference
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  dinf 0.1.1.dev155
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/RacimoLab/dinf"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/RacimoLab/dinf/issues/new?title=Issue%20on%20page%20%2F_modules/dinf/feature_extractor.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            
              <div>
                
  <h1>Source code for dinf.feature_extractor</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Mapping</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">numpy.typing</span> <span class="k">as</span> <span class="nn">npt</span>
<span class="kn">import</span> <span class="nn">tskit</span>

<span class="kn">from</span> <span class="nn">.vcf</span> <span class="kn">import</span> <span class="n">BagOfVcf</span>
<span class="kn">from</span> <span class="nn">.misc</span> <span class="kn">import</span> <span class="n">ts_ploidy_of_individuals</span><span class="p">,</span> <span class="n">ts_nodes_of_individuals</span>


<div class="viewcode-block" id="BinnedHaplotypeMatrix"><a class="viewcode-back" href="../../api.html#dinf.BinnedHaplotypeMatrix">[docs]</a><span class="k">class</span> <span class="nc">BinnedHaplotypeMatrix</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A factory for feature matrices of pseudo-haplotypes.</span>

<span class="sd">    The binned haplotype matrix is an :math:`n \\times m` matrix,</span>
<span class="sd">    where :math:`n` corresponds to the number of haplotypes</span>
<span class="sd">    (or polyploid genotypes, for unphased data) and :math:`m` corresponds to</span>
<span class="sd">    a set of equally sized bins along the sequence length. Each matrix entry</span>
<span class="sd">    contains the count of minor alleles in an individual&#39;s haplotype</span>
<span class="sd">    (or polyploid genotype) in a given bin.</span>

<span class="sd">    Only polymorphic SNPs are considered, and for multiallelic sites,</span>
<span class="sd">    only the first two alleles are used.</span>
<span class="sd">    Alleles are polarised by choosing the most frequent allele to be encoded</span>
<span class="sd">    as 0, and the other allele as 1. In the event that both alleles are equally</span>
<span class="sd">    frequent, the polarisation is chosen at random.</span>

<span class="sd">    As the features are intended to be passed to a covolutional neural network,</span>
<span class="sd">    the output dimensions are actually :math:`n \\times m \\times 1`, where the</span>
<span class="sd">    final dimension is the (unused) &quot;channels&quot; dimension for the convolution.</span>

<span class="sd">    Gower et al. 2021, https://doi.org/10.7554/eLife.64669</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">num_individuals</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">num_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">ploidy</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">maf_thresh</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
        <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param num_individuals:</span>
<span class="sd">            The number of individuals to include in the feature matrix.</span>
<span class="sd">        :param num_bins:</span>
<span class="sd">            The number of bins into which the sequence is partitioned.</span>
<span class="sd">            Each bin spans ``sequence_length / num_bins`` base pairs.</span>
<span class="sd">        :param ploidy:</span>
<span class="sd">            Ploidy of the individuals.</span>
<span class="sd">        :param phased:</span>
<span class="sd">            If True, the individuals&#39; haplotypes will each be included as</span>
<span class="sd">            independent rows in the feature matrix and the shape of the</span>
<span class="sd">            feature matrix will be ``(ploidy * num_individuals, num_bins, 1)``.</span>
<span class="sd">            If False, the allele counts for each individual will be summed</span>
<span class="sd">            across their chromosome copies and the shape of the feature matrix</span>
<span class="sd">            will be ``(num_individuals, num_bins, 1)``.</span>
<span class="sd">        :param maf_thresh:</span>
<span class="sd">            Minor allele frequency (MAF) threshold. Sites with MAF lower than</span>
<span class="sd">            this value are ignored.</span>
<span class="sd">        :param dtype:</span>
<span class="sd">            The numpy data type of the feature matrix. To save memory, we use</span>
<span class="sd">            an np.int8 by default, which assumes that counts are small.</span>
<span class="sd">            However, counts may overflow int8 for very large values of</span>
<span class="sd">            ``mu * Ne * num_individuals * sequence_length / num_bins``,</span>
<span class="sd">            in which case np.int16 might be preferred. Such cases have</span>
<span class="sd">            not been tested, but it&#39;s likely that increasing ``num_bins``</span>
<span class="sd">            is a better solution.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">num_individuals</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;must have num_individuals &gt;= 1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_bins</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;must have num_bins &gt;= 1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">maf_thresh</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">maf_thresh</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;must have 0 &lt;= maf_thresh &lt;= 1&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">dtype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;dtype must be np.int8, np.int16, or np.in32&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span> <span class="o">=</span> <span class="n">num_individuals</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span> <span class="o">=</span> <span class="n">num_bins</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_phased</span> <span class="o">=</span> <span class="n">phased</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_haplotypes</span> <span class="o">=</span> <span class="n">num_individuals</span> <span class="o">*</span> <span class="n">ploidy</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_dtype</span> <span class="o">=</span> <span class="n">dtype</span>
        <span class="c1"># We use a minimum threshold of 1 to exclude invariant sites.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_allele_count_threshold</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">maf_thresh</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_haplotypes</span><span class="p">)</span>

        <span class="c1"># The number of rows in the feature matrix is determined by</span>
        <span class="c1"># whether or not the data should be treated as phased.</span>
        <span class="c1"># If the data are not phased, we sum genotypes across the</span>
        <span class="c1"># chromosome copies. We call the rows &quot;pseudo haplotypes&quot;,</span>
        <span class="c1"># regardless of the phasing.</span>
        <span class="k">if</span> <span class="n">phased</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_num_pseudo_haplotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_haplotypes</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_num_pseudo_haplotypes</span> <span class="o">=</span> <span class="n">num_individuals</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_pseudo_haplotypes</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;must have at least two pseudo-haplotypes&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">shape</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Shape of the feature matrix.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_pseudo_haplotypes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_from_genotype_matrix</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">G</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">positions</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a pseudo-genotype matrix from a regular genotype matrix.</span>

<span class="sd">        Missing genotypes in the input are assumed to take the value -1,</span>
<span class="sd">        and the first allele has genotype 0, second allele genotype 1, etc.</span>
<span class="sd">        We consider only allele 0 and 1. For multiallelic sites this means</span>
<span class="sd">        all but the first two alleles are ignored.</span>

<span class="sd">        :param G:</span>
<span class="sd">            Genotype matrix with shape (num_sites, num_individuals, ploidy).</span>
<span class="sd">            The genotypes may be phased or unphased.</span>
<span class="sd">        :param positions:</span>
<span class="sd">            Vector of variant positions.</span>
<span class="sd">        :param sequence_length:</span>
<span class="sd">            The length of the sequence from which the matrix is derived.</span>
<span class="sd">        :param rng:</span>
<span class="sd">            Numpy random number generator. Used to randomly polarise alleles</span>
<span class="sd">            when there are multiple alleles with the same frequency.</span>
<span class="sd">        :return:</span>
<span class="sd">            Array with shape ``(num_pseudo_haplotypes, num_bins, 1)``.</span>
<span class="sd">            For a matrix :math:`M`, the :math:`M[i][j][0]`&#39;th entry is the</span>
<span class="sd">            count of minor alleles in the :math:`j`&#39;th bin of psdeudo-haplotype</span>
<span class="sd">            :math:`i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
        <span class="n">M</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_dtype</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">M</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">floor_divide</span><span class="p">(</span><span class="n">positions</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">int32</span>
        <span class="p">)</span>

        <span class="n">G_sites</span><span class="p">,</span> <span class="n">G_individuals</span><span class="p">,</span> <span class="n">G_ploidy</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">shape</span>
        <span class="k">assert</span> <span class="n">G_individuals</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span>
        <span class="k">assert</span> <span class="n">G_ploidy</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">,</span> <span class="p">(</span><span class="n">G_ploidy</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="p">(</span><span class="n">G_sites</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>

        <span class="c1"># Identify genotypes that aren&#39;t 0 or 1. These will be ignored later.</span>
        <span class="n">missing</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">G</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">ac0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ac1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">ac0</span><span class="p">,</span> <span class="n">ac1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_allele_count_threshold</span>
        <span class="c1"># Polarise 0 and 1 in genotype matrix by major allele frequency.</span>
        <span class="n">flip</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span>
            <span class="n">ac1</span> <span class="o">&gt;</span> <span class="n">ac0</span><span class="p">,</span>
            <span class="c1"># If allele counts are the same, randomly assign major allele.</span>
            <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">ac1</span> <span class="o">==</span> <span class="n">ac0</span><span class="p">,</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="n">G</span> <span class="o">^=</span> <span class="n">np</span><span class="o">.</span><span class="n">expand_dims</span><span class="p">(</span><span class="n">flip</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Exclude missing genotypes from the sums below.</span>
        <span class="n">G</span><span class="p">[</span><span class="n">missing</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_phased</span><span class="p">:</span>
            <span class="c1"># Collapse each individual&#39;s chromosome copies by summing counts.</span>
            <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">))</span>
            <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">genotypes</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">bins</span><span class="p">[</span><span class="n">keep</span><span class="p">],</span> <span class="n">G</span><span class="p">[</span><span class="n">keep</span><span class="p">]):</span>
            <span class="n">M</span><span class="p">[:,</span> <span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="n">genotypes</span>

        <span class="k">return</span> <span class="n">M</span>

<div class="viewcode-block" id="BinnedHaplotypeMatrix.from_ts"><a class="viewcode-back" href="../../api.html#dinf.BinnedHaplotypeMatrix.from_ts">[docs]</a>    <span class="k">def</span> <span class="nf">from_ts</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">ts</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TreeSequence</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a pseudo-genotype matrix from a tree sequence.</span>

<span class="sd">        :param ts: The tree sequence.</span>
<span class="sd">        :param numpy.random.Generator rng: Numpy random number generator.</span>
<span class="sd">        :return:</span>
<span class="sd">            Array with shape ``(num_pseudo_haplotypes, num_bins, 1)``.</span>
<span class="sd">            For a matrix :math:`M`, the :math:`M[i][j][0]`&#39;th entry is the</span>
<span class="sd">            count of minor alleles in the :math:`j`&#39;th bin of psdeudo-haplotype</span>
<span class="sd">            :math:`i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">ts</span><span class="o">.</span><span class="n">num_samples</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_haplotypes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Expected </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_haplotypes</span><span class="si">}</span><span class="s2"> haplotypes, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;but ts.num_samples == </span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">num_samples</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Sequence length is shorter than the number of bins&quot;</span><span class="p">)</span>

        <span class="n">G</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">genotype_matrix</span><span class="p">()</span>  <span class="c1"># shape is (num_sites, num_haplotypes)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">))</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_genotype_matrix</span><span class="p">(</span>
            <span class="n">G</span><span class="p">,</span> <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="BinnedHaplotypeMatrix.from_vcf"><a class="viewcode-back" href="../../api.html#dinf.BinnedHaplotypeMatrix.from_vcf">[docs]</a>    <span class="k">def</span> <span class="nf">from_vcf</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">vb</span><span class="p">:</span> <span class="n">BagOfVcf</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">min_seg_sites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a pseudo-genotype matrix from a region of a VCF/BCF.</span>

<span class="sd">        The genomic window is drawn uniformly at random from the sequences</span>
<span class="sd">        defined in the given :class:`BagOfVcf`.</span>

<span class="sd">        Individuals in the VCFs are sampled (without replacement) for</span>
<span class="sd">        inclusion in the output matrix. The size of the feature space</span>
<span class="sd">        can therefore be vastly increased by having more individuals</span>
<span class="sd">        in the VCFs than are needed for the feature dimensions.</span>

<span class="sd">        :param vb:</span>
<span class="sd">            The BagOfVcf object that describes the VCF/BCF files.</span>
<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the genomic window to be sampled.</span>
<span class="sd">        :param max_missing_genotypes:</span>
<span class="sd">            Consider only sites with fewer missing genotype calls than</span>
<span class="sd">            this number.</span>
<span class="sd">        :param min_seg_sites:</span>
<span class="sd">            Sampled genotype matrix must have at least this many variable</span>
<span class="sd">            sites (after filtering sites for missingness).</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            Numpy random number generator.</span>
<span class="sd">        :return:</span>
<span class="sd">            Array with shape ``(num_pseudo_haplotypes, num_bins, 1)``.</span>
<span class="sd">            For a matrix :math:`M`, the :math:`M[i][j][0]`&#39;th entry is the</span>
<span class="sd">            count of minor alleles in the :math:`j`&#39;th bin of psdeudo-haplotype</span>
<span class="sd">            :math:`i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">positions</span> <span class="o">=</span> <span class="n">vb</span><span class="o">.</span><span class="n">sample_genotype_matrix</span><span class="p">(</span>
            <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
            <span class="n">max_missing_genotypes</span><span class="o">=</span><span class="n">max_missing_genotypes</span><span class="p">,</span>
            <span class="n">min_seg_sites</span><span class="o">=</span><span class="n">min_seg_sites</span><span class="p">,</span>
            <span class="n">require_phased</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_phased</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">G_individuals</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">G_individuals</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Expected at least </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="si">}</span><span class="s2"> individuals in &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;the vcf bag, but only found </span><span class="si">{</span><span class="n">G_individuals</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Subsample individuals.</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">G_individuals</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">,</span> <span class="p">:]</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">2</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Mismatched ploidy among individuals.&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_from_genotype_matrix</span><span class="p">(</span>
            <span class="n">G</span><span class="p">,</span> <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span>
        <span class="p">)</span></div></div>


<div class="viewcode-block" id="MultipleBinnedHaplotypeMatrices"><a class="viewcode-back" href="../../api.html#dinf.MultipleBinnedHaplotypeMatrices">[docs]</a><span class="k">class</span> <span class="nc">MultipleBinnedHaplotypeMatrices</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A factory for labelled collections of :class:`BinnedHaplotypeMatrix` objects.</span>

<span class="sd">    One feature matrix is produced for each label. Labels typically</span>
<span class="sd">    correspond to populations, but this need not be the case.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">num_individuals</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
        <span class="n">num_bins</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
        <span class="n">ploidy</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span>
        <span class="c1"># TODO: label-specific options for phased and maf_thresh.</span>
        <span class="c1"># phased: dict,</span>
        <span class="c1"># maf_thresh: dict,</span>
        <span class="n">global_phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">global_maf_thresh</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param num_individuals:</span>
<span class="sd">            A dict that maps labels to the the number of individuals</span>
<span class="sd">            in the feature matrix.</span>
<span class="sd">        :param num_bins:</span>
<span class="sd">            A dict that maps labels to the number of bins into which the</span>
<span class="sd">            sequence is partitioned.</span>
<span class="sd">        :param ploidy:</span>
<span class="sd">            A dict that maps labels to the ploidy of the individuals.</span>
<span class="sd">        :param global_phased:</span>
<span class="sd">            If True, the individuals&#39; haplotypes will each be included as</span>
<span class="sd">            independent rows in each feature matrix and the shape of the</span>
<span class="sd">            feature matrix for label `l` will be</span>
<span class="sd">            ``(ploidy[l] * num_individuals[l], num_bins[l], 1)``.</span>
<span class="sd">            If False, the allele counts for each individual will be summed</span>
<span class="sd">            across their chromosome copies and the shape of the feature matrix</span>
<span class="sd">            will be ``(num_individuals[l], num_bins[l], 1)``.</span>
<span class="sd">        :param global_maf_thresh:</span>
<span class="sd">            Minor allele frequency (MAF) threshold. Sites with MAF lower than</span>
<span class="sd">            this value are ignored.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dict_args</span> <span class="o">=</span> <span class="p">[</span><span class="n">num_individuals</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">ploidy</span><span class="p">]</span>
        <span class="n">dict_strs</span> <span class="o">=</span> <span class="s2">&quot;num_individuals, num_bins, ploidy&quot;</span>
        <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">dict_args</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected dict for each of: </span><span class="si">{</span><span class="n">dict_strs</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">keys_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">dict_args</span><span class="p">]</span>
        <span class="n">keys</span> <span class="o">=</span> <span class="n">keys_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">keys</span> <span class="o">!=</span> <span class="n">other</span> <span class="k">for</span> <span class="n">other</span> <span class="ow">in</span> <span class="n">keys_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Must use the same dict keys for each of: </span><span class="si">{</span><span class="n">dict_strs</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">label</span><span class="p">:</span> <span class="n">BinnedHaplotypeMatrix</span><span class="p">(</span>
                <span class="n">num_individuals</span><span class="o">=</span><span class="n">num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">],</span>
                <span class="n">num_bins</span><span class="o">=</span><span class="n">num_bins</span><span class="p">[</span><span class="n">label</span><span class="p">],</span>
                <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">],</span>
                <span class="n">phased</span><span class="o">=</span><span class="n">global_phased</span><span class="p">,</span>
                <span class="n">maf_thresh</span><span class="o">=</span><span class="n">global_maf_thresh</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">keys</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span> <span class="o">=</span> <span class="n">num_individuals</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span> <span class="o">=</span> <span class="n">num_bins</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span> <span class="o">=</span> <span class="n">ploidy</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_global_phased</span> <span class="o">=</span> <span class="n">global_phased</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_global_maf_thresh</span> <span class="o">=</span> <span class="n">global_maf_thresh</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">shape</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;Shape of the feature matrix.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">label</span><span class="p">:</span> <span class="n">bhm</span><span class="o">.</span><span class="n">shape</span> <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">bhm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<div class="viewcode-block" id="MultipleBinnedHaplotypeMatrices.from_ts"><a class="viewcode-back" href="../../api.html#dinf.MultipleBinnedHaplotypeMatrices.from_ts">[docs]</a>    <span class="k">def</span> <span class="nf">from_ts</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">ts</span><span class="p">:</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TreeSequence</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
        <span class="n">individuals</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">npt</span><span class="o">.</span><span class="n">NDArray</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">integer</span><span class="p">]],</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create pseudo-genotype matrices from a tree sequence.</span>

<span class="sd">        :param ts: The tree sequence.</span>
<span class="sd">        :param numpy.random.Generator rng: Numpy random number generator.</span>
<span class="sd">        :param individuals:</span>
<span class="sd">            A mapping from label to an array of individuals.</span>
<span class="sd">        :return:</span>
<span class="sd">            A dictionary mapping a label ``l`` to a feature array.</span>
<span class="sd">            Each array has shape ``(num_pseudo_haplotypes[l], num_bins[l], 1)``.</span>
<span class="sd">            For an array :math:`M`, the :math:`M[i][j][0]`&#39;th entry is the</span>
<span class="sd">            count of minor alleles in the :math:`j`&#39;th bin of psdeudo-haplotype</span>
<span class="sd">            :math:`i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">individuals</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Labels of individuals </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span><span class="si">}</span><span class="s2"> don&#39;t match &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;feature labels </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">genotype_matrix</span><span class="p">()</span>  <span class="c1"># shape is (num_sites, num_haplotypes)</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
        <span class="n">labelled_features</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">l_individuals</span> <span class="ow">in</span> <span class="n">individuals</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span><span class="p">[</span><span class="n">label</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: sequence length (</span><span class="si">{</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2">) is &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;shorter than the number of bins (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_bins</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="si">}</span><span class="s2">).&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">l_individuals</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: expected </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;individuals, but got </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">l_individuals</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
            <span class="n">ploidy</span> <span class="o">=</span> <span class="n">ts_ploidy_of_individuals</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">l_individuals</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">ploidy</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">]):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: not all individuals have ploidy == </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> ploidies: </span><span class="si">{</span><span class="n">ploidy</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>
            <span class="n">nodes</span> <span class="o">=</span> <span class="n">ts_nodes_of_individuals</span><span class="p">(</span><span class="n">ts</span><span class="p">,</span> <span class="n">l_individuals</span><span class="p">)</span>
            <span class="n">H</span> <span class="o">=</span> <span class="n">G</span><span class="p">[:,</span> <span class="n">nodes</span><span class="p">]</span>
            <span class="n">H</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">H</span><span class="p">,</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">]))</span>
            <span class="n">labelled_features</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">_from_genotype_matrix</span><span class="p">(</span>
                <span class="n">H</span><span class="p">,</span> <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="n">ts</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">labelled_features</span></div>

<div class="viewcode-block" id="MultipleBinnedHaplotypeMatrices.from_vcf"><a class="viewcode-back" href="../../api.html#dinf.MultipleBinnedHaplotypeMatrices.from_vcf">[docs]</a>    <span class="k">def</span> <span class="nf">from_vcf</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">vb</span><span class="p">:</span> <span class="n">BagOfVcf</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">min_seg_sites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create pseudo-genotype matrices from a region of a VCF/BCF.</span>

<span class="sd">        The genomic window is drawn uniformly at random from the sequences</span>
<span class="sd">        defined in the given :class:`BagOfVcf`.</span>

<span class="sd">        Individuals in the VCFs are sampled (without replacement) for</span>
<span class="sd">        inclusion in the output matrix. The size of the feature space</span>
<span class="sd">        can therefore be vastly increased by having more individuals</span>
<span class="sd">        in the VCFs than are needed for the feature dimensions.</span>

<span class="sd">        :param vb:</span>
<span class="sd">            The BagOfVcf object that describes the VCF/BCF files.</span>
<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the genomic window to be sampled.</span>
<span class="sd">        :param max_missing_genotypes:</span>
<span class="sd">            Consider only sites with fewer missing genotype calls than</span>
<span class="sd">            this number.</span>
<span class="sd">        :param min_seg_sites:</span>
<span class="sd">            Sampled genotype matrix must have at least this many variable</span>
<span class="sd">            sites (after filtering sites for missingness).</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            Numpy random number generator.</span>
<span class="sd">        :return:</span>
<span class="sd">            A dictionary mapping a label ``l`` to a feature array.</span>
<span class="sd">            Each array has shape ``(num_pseudo_haplotypes[l], num_bins[l], 1)``.</span>
<span class="sd">            For an array :math:`M`, the :math:`M[i][j][0]`&#39;th entry is the</span>
<span class="sd">            count of minor alleles in the :math:`j`&#39;th bin of psdeudo-haplotype</span>
<span class="sd">            :math:`i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">vb</span><span class="o">.</span><span class="n">samples</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">!=</span> <span class="n">vb</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">sample_labels</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="n">vb</span><span class="o">.</span><span class="n">samples</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="nb">list</span><span class="p">(</span><span class="n">vb</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Feature labels </span><span class="si">{</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="p">)</span><span class="si">}</span><span class="s2"> don&#39;t match the &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;vcf bag&#39;s sample labels: </span><span class="si">{</span><span class="n">sample_labels</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="n">G</span><span class="p">,</span> <span class="n">positions</span> <span class="o">=</span> <span class="n">vb</span><span class="o">.</span><span class="n">sample_genotype_matrix</span><span class="p">(</span>
            <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span>
            <span class="n">max_missing_genotypes</span><span class="o">=</span><span class="n">max_missing_genotypes</span><span class="p">,</span>
            <span class="n">min_seg_sites</span><span class="o">=</span><span class="n">min_seg_sites</span><span class="p">,</span>
            <span class="n">require_phased</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_global_phased</span><span class="p">,</span>
            <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">num_samples</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">vb</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
        <span class="n">offsets</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">num_samples</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">labelled_features</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">bh_matrix</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bh_matrices</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="k">if</span> <span class="n">num_samples</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: Expected at least </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;individuals in the vcf bag, but only found </span><span class="si">{</span><span class="n">num_samples</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>

            <span class="c1"># Subsample individuals.</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">offsets</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                <span class="n">num_samples</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_num_individuals</span><span class="p">[</span><span class="n">label</span><span class="p">],</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>
            <span class="n">H</span> <span class="o">=</span> <span class="n">G</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">,</span> <span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">]]</span>
            <span class="n">ploidy_pad</span> <span class="o">=</span> <span class="n">G</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ploidy</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="p">:]</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">H</span> <span class="o">==</span> <span class="o">-</span><span class="mi">2</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">ploidy_pad</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">2</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: mismatched ploidy among individuals.&quot;</span><span class="p">)</span>
            <span class="n">labelled_features</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">bh_matrix</span><span class="o">.</span><span class="n">_from_genotype_matrix</span><span class="p">(</span>
                <span class="n">H</span><span class="p">,</span> <span class="n">positions</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span> <span class="n">sequence_length</span><span class="o">=</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="o">=</span><span class="n">rng</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">labelled_features</span></div></div>
</pre></div>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Graham Gower<br/>
    
        &copy; Copyright 2021.<br/>
      <div class="extra_footer">
        dinf 0.1.1.dev155+g381d22e
      </div>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>