
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dinf.vcf &#8212; Dinf</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">Dinf</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../introduction.html">
                    Introduction
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Getting started
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../installation.html">
   Installation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorial
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/creating_a_model.html">
   1. Creating a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/testing_a_model.html">
   2. Testing a Dinf model
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/abc.html">
   3. An ABC analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Topic guides
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/accuracy.html">
   Improving discriminator accuracy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/features.html">
   Feature matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/performance.html">
   Performance characteristics
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/multiple_demes.html">
   Models with multiple demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/empirical_data.html">
   Empirical data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/abc-gan.html">
   ABC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/mcmc-gan.html">
   MCMC-GAN
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../guide/pg-gan.html">
   PG-GAN (simulated annealing)
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../api.html">
   API reference
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../cli.html">
   CLI reference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellaneous
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../development.html">
   Development
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../CHANGELOG.html">
   Changelog
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            dinf 0.2.1.dev12
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/RacimoLab/dinf"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/RacimoLab/dinf/issues/new?title=Issue%20on%20page%20%2F_modules/dinf/vcf.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <h1>Source code for dinf.vcf</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">cyvcf2</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="get_samples_from_1kgp_metadata"><a class="viewcode-back" href="../../api.html#dinf.get_samples_from_1kgp_metadata">[docs]</a><span class="k">def</span> <span class="nf">get_samples_from_1kgp_metadata</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">populations</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get sample IDs for 1000 Genomes Project populations.</span>
<span class="sd">    Related individuals are removed based on the FatherID and MotherID.</span>

<span class="sd">    :param filename:</span>
<span class="sd">        Path to the file &quot;20130606_g1k_3202_samples_ped_population.txt&quot;</span>
<span class="sd">        This file can be downloaded from</span>
<span class="sd">        http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/</span>
<span class="sd">    :param populations:</span>
<span class="sd">        List of populations to extract.</span>
<span class="sd">    :return:</span>
<span class="sd">        A dict mapping population names to a list of sample IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">recfromtxt</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>
    <span class="c1"># Remove related individuals.</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">FatherID</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">]</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MotherID</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">data</span><span class="o">.</span><span class="n">SampleID</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">Population</span> <span class="o">==</span> <span class="n">pop</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">}</span></div>


<div class="viewcode-block" id="get_contig_lengths"><a class="viewcode-back" href="../../api.html#dinf.get_contig_lengths">[docs]</a><span class="k">def</span> <span class="nf">get_contig_lengths</span><span class="p">(</span>
    <span class="n">filename</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="n">keep_contigs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load contig lengths from a whitespace-separated file like an fai (fasta index).</span>

<span class="sd">    The file must have (at least) two columns, the first specifies the</span>
<span class="sd">    contig ID, and the second is the contig length. Additional columns</span>
<span class="sd">    are ignored.</span>

<span class="sd">    :param filename: The path to the file.</span>
<span class="sd">    :param keep_contigs: Use only the specified contigs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">keep_contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">keep_contigs</span><span class="p">)</span>
        <span class="n">seen_contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">maxsplit</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span>
            <span class="n">contig</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seen_contigs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">contig</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">contig</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">keep_contigs</span><span class="p">:</span>
                    <span class="k">continue</span>
            <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">data</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span> <span class="o">=</span> <span class="n">length</span>
    <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">contigs_not_found</span> <span class="o">=</span> <span class="n">keep_contigs</span> <span class="o">-</span> <span class="n">seen_contigs</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Requested contigs not found: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span></div>


<span class="k">def</span> <span class="nf">filter_missingness</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">max_missing_genotypes</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="n">max_missing_genotypes</span>


<span class="k">def</span> <span class="nf">filter_invariant</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
    <span class="n">num_non_missing</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac0</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac1</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac2</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac3</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac0</span><span class="p">,</span> <span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac1</span><span class="p">),</span>
        <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac2</span><span class="p">,</span> <span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac3</span><span class="p">),</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">get_genotype_matrix</span><span class="p">(</span>
    <span class="n">vcf</span><span class="p">:</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">,</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">require_phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a genotype matrix for the specified genomic window.</span>

<span class="sd">    The input region &#39;chrom:start-end&#39; uses 1-based closed coordinates,</span>
<span class="sd">    to match htslib (upon which cyvcf2 is built).</span>
<span class="sd">    The returned matrix has shape (num_sites, num_individuals, ploidy),</span>
<span class="sd">    where each site is a segregating SNP.</span>

<span class="sd">    :param vcf:</span>
<span class="sd">        A cyvcf2.VCF instance from which the genotypes will be obtained.</span>
<span class="sd">    :param chrom:</span>
<span class="sd">        Chromosome name (or non-chromosome sequence name).</span>
<span class="sd">    :param start:</span>
<span class="sd">        1-based start coordinate of the genomic window.</span>
<span class="sd">    :param end:</span>
<span class="sd">        1-based end coordinate of the genomic window.</span>
<span class="sd">    :param max_missing_genotypes:</span>
<span class="sd">        Only retain sites with at most this many missing genotype calls.</span>
<span class="sd">    :param require_phased:</span>
<span class="sd">        If True, raise an error if genotypes are not phased.</span>
<span class="sd">    :return:</span>
<span class="sd">        A 2-tuple of (G, positions) where</span>
<span class="sd">         - ``G`` is a (num_sites, num_individuals, ploidy) genotype matrix,</span>
<span class="sd">         - ``positions`` are the site coordinates, as a zero-based offset from</span>
<span class="sd">           the ``start`` coordinate.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gt_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">positions_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">):</span>
        <span class="c1"># XXX: Many variant fields are broken for ploidy &gt; 2.</span>
        <span class="c1"># https://github.com/brentp/cyvcf2/issues/227</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">variant</span><span class="o">.</span><span class="n">is_snp</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">a</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="n">a</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Check phasing. For ploidy == 1, genotypes are reported as unphased.</span>
        <span class="k">if</span> <span class="n">require_phased</span> <span class="ow">and</span> <span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">a</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unphased genotypes at </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="si">{</span><span class="n">variant</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">gt_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
        <span class="n">positions_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gt_list</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="c1"># Most likely an &quot;inhomogeneous shape&quot; error, caused by different</span>
        <span class="c1"># ploidies at different sites. Catch and reraise to provide</span>
        <span class="c1"># a more user-friendly error message.</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Mismatched ploidy among sites in </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

    <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">positions_list</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="n">start</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Filter for missingness.</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">filter_missingness</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">max_missing_genotypes</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Filter invariant sites.</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">filter_invariant</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">G</span><span class="p">,</span> <span class="n">positions</span>


<div class="viewcode-block" id="BagOfVcf"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf">[docs]</a><span class="k">class</span> <span class="nc">BagOfVcf</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A collection of indexed VCF or BCF files.</span>

<span class="sd">    VCF data are sometimes contained in a single file, and sometimes split into</span>
<span class="sd">    multiple files by chromosome. To remove the burden of dealing with both</span>
<span class="sd">    of these common cases, this class maps a contig ID to a :class:`cyvcf2.VCF`</span>
<span class="sd">    object for that contig. The class implements Python&#39;s :term:`python:mapping`</span>
<span class="sd">    protocol, plus methods for sampling regions of the genome at random.</span>

<span class="sd">    .. code::</span>

<span class="sd">        import dinf</span>

<span class="sd">        vcfs = dinf.BagOfVcf([&quot;chr1.bcf&quot;, &quot;chr2.bcf&quot;, &quot;chr3.bcf&quot;])</span>

<span class="sd">        # Iterate over contig IDs.</span>
<span class="sd">        assert len(vcfs) == 3</span>
<span class="sd">        assert list(vcfs) == [&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr3&quot;]</span>

<span class="sd">        # Lookup a cyvcf2.VCF object by contig ID.</span>
<span class="sd">        chr1 = vcfs[&quot;chr1&quot;]</span>
<span class="sd">        assert chr1.fname == &quot;chr1.bcf&quot;</span>
<span class="sd">        first_variant = next(chr1)</span>
<span class="sd">        assert first_variant.CHROM == &quot;chr1&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">lengths</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Lengths of the contigs in the bag.</span>

<span class="sd">    The order matches the contig order obtained when iterating.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">samples</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">|</span> <span class="kc">None</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A dictionary that maps a label to a list of individual names.</span>

<span class="sd">    The individual names correspond to the VCF columns for which genotypes</span>
<span class="sd">    will be sampled. This is a bookkeeping device that records which genotypes</span>
<span class="sd">    belong to which label (e.g. which population). If None, it is assumed</span>
<span class="sd">    that all individuals in the VCF will be treated as exchangeable.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">files</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="o">/</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">samples</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">contig_lengths</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param files:</span>
<span class="sd">            An iterable of filenames. Each file must be an indexed VCF or BCF,</span>
<span class="sd">            and must contain the FORMAT/GT field.</span>
<span class="sd">        :param contig_lengths:</span>
<span class="sd">            A dict mapping a contig name to contig length. Only the contigs in</span>
<span class="sd">            this dict will be used.</span>
<span class="sd">        :param samples:</span>
<span class="sd">            A dictionary that maps a label to a list of individual names,</span>
<span class="sd">            where the individual names correspond to the VCF columns</span>
<span class="sd">            for which genotypes will be sampled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">individuals</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="n">samples</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="c1"># Silence some warnings from cyvcf2.</span>
        <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
            <span class="c1"># We check if a contig is usable for a given vcf by querying</span>
            <span class="c1"># that contig for variants. If there are no variants (e.g. vcfs are</span>
            <span class="c1"># split by chromosome so each vcf has data for only one contig),</span>
            <span class="c1"># then cyvcf2 warns us.</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
                <span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;no intervals found&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span>
            <span class="p">)</span>
            <span class="c1"># We detect this condition and raise our own error below.</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
                <span class="s2">&quot;ignore&quot;</span><span class="p">,</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;not all requested samples found&quot;</span><span class="p">,</span>
                <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_fill_bag</span><span class="p">(</span>
                <span class="n">files</span><span class="o">=</span><span class="n">files</span><span class="p">,</span> <span class="n">contig_lengths</span><span class="o">=</span><span class="n">contig_lengths</span><span class="p">,</span> <span class="n">individuals</span><span class="o">=</span><span class="n">individuals</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">samples</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">_fill_bag</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">files</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="n">contig_lengths</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">individuals</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a mapping from contig ID to :class:`cyvcf2.VCF` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">contig2file</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">contig2vcf</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;File list contains duplicates.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">individuals</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Individuals list contains duplicates.&quot;</span><span class="p">)</span>

        <span class="n">all_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">all_samples_file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">contig_lengths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">contig_lengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">contigs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">contig_lengths</span><span class="p">)</span>
        <span class="n">contigs_seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">vcf</span> <span class="o">=</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">individuals</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;GT&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> doesn&#39;t contain GT field.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
                <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.tbi&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
                <span class="ow">or</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.csi&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
            <span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No index found for </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">all_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">all_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
                    <span class="n">all_samples_file</span> <span class="o">=</span> <span class="n">file</span>
                <span class="k">elif</span> <span class="n">all_samples</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> has different samples than </span><span class="si">{</span><span class="n">all_samples_file</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">individuals_not_found</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Requested individuals not found in </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">: &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">individuals_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
            <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">vcf</span><span class="o">.</span><span class="n">seqlens</span>
                <span class="k">except</span> <span class="ne">AttributeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> doesn&#39;t contain contig lengths. &quot;</span>
                        <span class="s2">&quot;You must provide a contig_lengths argument.&quot;</span>
                    <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">seqnames</span><span class="p">):</span>
                <span class="n">contigs_seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">contig_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">contig_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="c1"># Check there&#39;s at least one variant. If present, we assume</span>
                <span class="c1"># there exist usable variants (i.e. SNPs) for the contig.</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">next</span><span class="p">(</span><span class="n">vcf</span><span class="p">(</span><span class="n">contig_id</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="n">contig2file</span><span class="p">:</span>
                    <span class="n">first_file</span> <span class="o">=</span> <span class="n">contig2file</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Both </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">first_file</span><span class="si">}</span><span class="s2"> contain records for &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;sequence &#39;</span><span class="si">{</span><span class="n">contig_id</span><span class="si">}</span><span class="s2">&#39;.&quot;</span>
                    <span class="p">)</span>
                <span class="n">contig2file</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">file</span>
                <span class="n">contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcf</span>
                <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">contig_lengths</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">seqlens</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>  <span class="c1"># type: ignore[index]</span>

        <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">contigs_not_found</span> <span class="o">=</span> <span class="n">contigs</span> <span class="o">-</span> <span class="n">contigs_seen</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Requested contigs not found: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contig2file</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No usable vcf/bcf files in the list.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">all_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_samples</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_individuals</span> <span class="o">=</span> <span class="n">individuals</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span> <span class="o">=</span> <span class="n">contig2file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span> <span class="o">=</span> <span class="n">contig2vcf</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromiter</span><span class="p">(</span><span class="n">contig_lengths</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

        <span class="c1"># The VCF objects are not valid in child processes, so we set a flag</span>
        <span class="c1"># before forking to ensure they&#39;re reopened when used in the child.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">os</span><span class="o">.</span><span class="n">register_at_fork</span><span class="p">(</span><span class="n">before</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_close</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterate over contig IDs in the bag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">yield from</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the :class:`cyvcf2.VCF` containing the given contig.</span>

<span class="sd">        :param contig_id:</span>
<span class="sd">            The contig ID.</span>
<span class="sd">        :return:</span>
<span class="sd">            The :class:`cyvcf2.VCF` object.</span>
<span class="sd">        :rtype: cyvcf2.VCF</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">contig_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;key must be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reopen</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The number of contigs in the mapping.</span>

<span class="sd">        :return:</span>
<span class="sd">            The size of the bag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">_reopen</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">file2vcf</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">file</span><span class="p">:</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_individuals</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">contig_id</span><span class="p">:</span> <span class="n">file2vcf</span><span class="p">[</span><span class="n">file</span><span class="p">]</span> <span class="k">for</span> <span class="n">contig_id</span><span class="p">,</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">False</span>

<div class="viewcode-block" id="BagOfVcf.sample_regions"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf.sample_regions">[docs]</a>    <span class="k">def</span> <span class="nf">sample_regions</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sample a list of (chrom, start, end) triplets.</span>

<span class="sd">        :param size:</span>
<span class="sd">            Number of genomic windows to sample.</span>
<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the sequence to sample.</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            The numpy random number generator.</span>
<span class="sd">        :return:</span>
<span class="sd">            List of 3-tuples: (chrom, start, end).</span>
<span class="sd">            The start and end coordinates are 1-based and inclusive, to match</span>
<span class="sd">            the usual convention for &#39;chrom:start-end&#39;, e.g. in bcftools.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">long_enough</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">&gt;=</span> <span class="n">sequence_length</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">long_enough</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No contigs with length &gt;= </span><span class="si">{</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">contig_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)[</span><span class="n">long_enough</span><span class="p">]</span>
        <span class="n">contig_lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">long_enough</span><span class="p">]</span>

        <span class="n">p</span> <span class="o">=</span> <span class="n">contig_lengths</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">contig_lengths</span><span class="p">)</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">contig_ids</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">p</span><span class="p">)</span>
        <span class="n">upper_limit</span> <span class="o">=</span> <span class="p">(</span><span class="n">contig_lengths</span> <span class="o">-</span> <span class="n">sequence_length</span><span class="p">)[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="n">upper_limit</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">endpoint</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">sequence_length</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">contig_ids</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">regions</span></div>

<div class="viewcode-block" id="BagOfVcf.sample_genotype_matrix"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf.sample_genotype_matrix">[docs]</a>    <span class="k">def</span> <span class="nf">sample_genotype_matrix</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">min_seg_sites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">require_phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
        <span class="n">retries</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sample a genotype matrix uniformly at random from the genome.</span>

<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the sequence to sample.</span>
<span class="sd">        :param max_missing_genotypes:</span>
<span class="sd">            Consider only sites with at most this many missing genotype calls.</span>
<span class="sd">        :param min_seg_sites:</span>
<span class="sd">            Sampled genotype matrix must have at least this many variable</span>
<span class="sd">            sites (after filtering sites for missingness).</span>
<span class="sd">        :param require_phased:</span>
<span class="sd">            If True, raise an error if genotypes are not phased.</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            Numpy random number generator.</span>
<span class="sd">        :param retries:</span>
<span class="sd">            Maximum number of attempts allowed to find a genomic window with</span>
<span class="sd">            the required number of segregating sites.</span>
<span class="sd">        :return:</span>
<span class="sd">            A 2-tuple of (G, positions) where</span>
<span class="sd">             - ``G`` is a (num_sites, num_individuals, ploidy) genotype matrix,</span>
<span class="sd">             - ``positions`` are the site coordinates, as an offset from the</span>
<span class="sd">               the start of the genomic window.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">retries</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">retries</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_regions</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="p">)</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="n">G</span><span class="p">,</span> <span class="n">positions</span> <span class="o">=</span> <span class="n">get_genotype_matrix</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">[</span><span class="n">chrom</span><span class="p">],</span>
                <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
                <span class="n">max_missing_genotypes</span><span class="o">=</span><span class="n">max_missing_genotypes</span><span class="p">,</span>
                <span class="n">require_phased</span><span class="o">=</span><span class="n">require_phased</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_seg_sites</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed to sample genotype matrix with at least </span><span class="si">{</span><span class="n">min_seg_sites</span><span class="si">}</span><span class="s2"> &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;segregating sites after </span><span class="si">{</span><span class="n">retries</span><span class="si">}</span><span class="s2"> attempts.&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">G</span><span class="p">,</span> <span class="n">positions</span></div></div>
</pre></div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Graham Gower<br/>
  
      &copy; Copyright 2021-2022.<br/>
    <div class="extra_footer">
      dinf 0.2.1.dev12+g4ba073b
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>