
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>dinf.vcf &#8212; Dinf</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/basic_mod.css?v=0.7.0-1" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=8bf782fb4ee92b3d3646425e50f299c4e1fd152d"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <script src="../../_static/js/petite-vue.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
  </head><body data-dark_mode_code_blocks="true">

<div id="top_nav">
    

    <nav>
        
            
        

        <p id="toggle_sidebar">
            <a href="#" title="Toggle sidebar">|||</a>
        </p>
        <h1><a href="../../introduction.html" title="Go to homepage">Dinf</a></h1>
            <a id="source_link" href="https://github.com/RacimoLab/dinf">
    
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512">
            <path fill="white" d="M 244.8,8 C 106.1,8 0,113.3 0,252 c 0,110.9 69.8,205.8 169.5,239.2 12.8,2.3 17.3,-5.6 17.3,-12.1 0,-6.2 -0.3,-40.4 -0.3,-61.4 0,0 -70,15 -84.7,-29.8 0,0 -11.4,-29.1 -27.8,-36.6 0,0 -22.9,-15.7 1.6,-15.4 0,0 24.9,2 38.6,25.8 21.9,38.6 58.6,27.5 72.9,20.9 2.3,-16 8.8,-27.1 16,-33.7 -55.9,-6.2 -112.3,-14.3 -112.3,-110.5 0,-27.5 7.6,-41.3 23.6,-58.9 -2.6,-6.5 -11.1,-33.3 2.6,-67.9 20.9,-6.5 69,27 69,27 20,-5.6 41.5,-8.5 62.8,-8.5 21.3,0 42.8,2.9 62.8,8.5 0,0 48.1,-33.6 69,-27 13.7,34.7 5.2,61.4 2.6,67.9 16,17.7 25.8,31.5 25.8,58.9 0,96.5 -58.9,104.2 -114.8,110.5 9.2,7.9 17,22.9 17,46.4 0,33.7 -0.3,75.4 -0.3,83.6 0,6.5 4.6,14.4 17.3,12.1 C 428.2,457.8 496,362.9 496,252 496,113.3 383.5,8 244.8,8 Z"/>
        </svg>
    
</a>
        

        <a id="mode_toggle" href="#" @click.prevent="handleClick" :title="mode">
    <template v-if="mode == 'light'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_light"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><circle cx="39.311" cy="39.524" r="15.734" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.212,4.901c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901l-0,9.614c-0,2.705 2.196,4.901 4.9,4.901c2.705,0 4.901,-2.196 4.901,-4.901l0,-9.614Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M67.48,18.073c1.913,-1.912 1.913,-5.018 0,-6.931c-1.912,-1.912 -5.018,-1.912 -6.931,0l-6.798,6.799c-1.912,1.912 -1.912,5.018 0,6.931c1.913,1.912 5.018,1.912 6.931,-0l6.798,-6.799Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.728,61.108c1.912,-1.913 1.912,-5.018 -0,-6.931c-1.913,-1.913 -5.019,-1.913 -6.931,-0l-6.799,6.798c-1.912,1.913 -1.912,5.019 0,6.931c1.913,1.913 5.019,1.913 6.931,0l6.799,-6.798Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.682,54.177c-1.913,-1.913 -5.018,-1.913 -6.931,-0c-1.912,1.913 -1.912,5.018 0,6.931l6.798,6.798c1.913,1.913 5.019,1.913 6.931,0c1.913,-1.912 1.913,-5.018 0,-6.931l-6.798,-6.798Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M4.901,34.623c-2.705,0 -4.901,2.196 -4.901,4.901c0,2.705 2.196,4.901 4.901,4.901l9.614,0c2.705,0 4.901,-2.196 4.901,-4.901c0,-2.705 -2.196,-4.901 -4.901,-4.901l-9.614,0Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.212,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901l-0,9.614c-0,2.705 2.196,4.901 4.9,4.901c2.705,-0 4.901,-2.196 4.901,-4.901l0,-9.614Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M18.929,11.142c-1.912,-1.912 -5.018,-1.912 -6.931,0c-1.912,1.913 -1.912,5.019 0,6.931l6.799,6.799c1.912,1.912 5.018,1.912 6.931,-0c1.912,-1.913 1.912,-5.019 -0,-6.931l-6.799,-6.799Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.108,34.623c-2.705,0 -4.901,2.196 -4.901,4.901c-0,2.705 2.196,4.901 4.901,4.901l9.614,0c2.705,0 4.901,-2.196 4.901,-4.901c-0,-2.705 -2.196,-4.901 -4.901,-4.901l-9.614,0Z" style="fill:#fff;"/></g></g></g></svg>
    </template>

    <template v-if="mode == 'dark'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_dark"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><circle cx="39.311" cy="39.524" r="15.734" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.212,14.515c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,0 -4.901,2.196 -4.901,4.901c0,2.705 2.197,4.901 4.901,4.901c2.705,0 4.901,-2.196 4.901,-4.901Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M60.662,24.892c1.902,-1.902 1.902,-4.99 0,-6.892l-0.04,-0.039c-1.901,-1.902 -4.989,-1.902 -6.891,-0c-1.901,1.901 -1.901,4.989 0,6.891l0.04,0.04c1.902,1.901 4.989,1.901 6.891,-0Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.732,61.103c1.91,-1.91 1.91,-5.011 0,-6.921l-0.009,-0.01c-1.91,-1.91 -5.012,-1.91 -6.921,-0c-1.91,1.91 -1.91,5.011 -0,6.921l0.01,0.01c1.909,1.91 5.011,1.91 6.92,-0Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.672,54.167c-1.907,-1.907 -5.004,-1.907 -6.911,0l-0.02,0.02c-1.907,1.907 -1.907,5.004 0,6.911c1.907,1.907 5.004,1.907 6.911,-0l0.02,-0.02c1.907,-1.907 1.907,-5.004 0,-6.911Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M14.52,34.623c-2.702,0 -4.896,2.194 -4.896,4.896l0,0.01c0,2.702 2.194,4.896 4.896,4.896c2.702,0 4.896,-2.194 4.896,-4.896l-0,-0.01c-0,-2.702 -2.194,-4.896 -4.896,-4.896Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.212,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.901,2.196 -4.901,4.901c0,2.704 2.197,4.9 4.901,4.9c2.705,0 4.901,-2.196 4.901,-4.9Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M25.73,17.943c-1.911,-1.911 -5.015,-1.911 -6.926,0l-0.005,0.005c-1.911,1.911 -1.911,5.015 0,6.926c1.911,1.911 5.015,1.911 6.926,0l0.005,-0.005c1.911,-1.911 1.911,-5.014 -0,-6.926Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.098,34.623c-2.699,0 -4.891,2.192 -4.891,4.892l-0,0.019c-0,2.699 2.192,4.891 4.891,4.891c2.7,0 4.892,-2.192 4.892,-4.891l0,-0.019c0,-2.7 -2.192,-4.892 -4.892,-4.892Z" style="fill:#fff;"/></g></g></g></svg>
    </template>

    <template v-if="mode == 'darkest'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_darkest"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><path d="M39.315,23.791c8.684,-0 15.734,7.05 15.734,15.733c0,8.684 -7.05,15.734 -15.734,15.734c-8.683,0 -15.733,-7.05 -15.733,-15.734c-0,-8.683 7.05,-15.733 15.733,-15.733Zm0,4.737c6.069,0 10.997,4.927 10.997,10.996c-0,6.069 -4.928,10.996 -10.997,10.996c-6.068,0 -10.996,-4.927 -10.996,-10.996c0,-6.069 4.928,-10.996 10.996,-10.996Z" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.216,14.515c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,0 -4.9,2.196 -4.9,4.901c-0,2.705 2.196,4.901 4.9,4.901c2.705,0 4.901,-2.196 4.901,-4.901Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M60.666,24.892c1.902,-1.902 1.902,-4.99 0,-6.892l-0.04,-0.039c-1.901,-1.902 -4.989,-1.902 -6.891,-0c-1.901,1.901 -1.901,4.989 0,6.891l0.04,0.04c1.902,1.901 4.99,1.901 6.891,-0Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.737,61.103c1.909,-1.91 1.909,-5.011 -0,-6.921l-0.01,-0.01c-1.91,-1.91 -5.011,-1.91 -6.921,-0c-1.91,1.91 -1.91,5.011 -0,6.921l0.01,0.01c1.91,1.91 5.011,1.91 6.921,-0Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.676,54.167c-1.907,-1.907 -5.004,-1.907 -6.911,0l-0.02,0.02c-1.907,1.907 -1.907,5.004 0,6.911c1.907,1.907 5.004,1.907 6.911,-0l0.02,-0.02c1.907,-1.907 1.907,-5.004 0,-6.911Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M14.524,34.623c-2.702,0 -4.896,2.194 -4.896,4.896l0,0.01c0,2.702 2.194,4.896 4.896,4.896c2.702,0 4.896,-2.194 4.896,-4.896l0,-0.01c0,-2.702 -2.194,-4.896 -4.896,-4.896Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.216,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901c-0,2.704 2.196,4.9 4.9,4.9c2.705,0 4.901,-2.196 4.901,-4.9Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M25.734,17.943c-1.911,-1.911 -5.015,-1.911 -6.926,0l-0.005,0.005c-1.911,1.911 -1.911,5.015 0,6.926c1.911,1.911 5.015,1.911 6.926,0l0.005,-0.005c1.911,-1.911 1.911,-5.014 0,-6.926Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.103,34.623c-2.7,0 -4.892,2.192 -4.892,4.892l-0,0.019c-0,2.699 2.192,4.891 4.892,4.891c2.699,0 4.891,-2.192 4.891,-4.891l0,-0.019c0,-2.7 -2.192,-4.892 -4.891,-4.892Z" style="fill:#fff;"/></g></g></g></svg>
    </template>
</a>

<script>
(function() {
    const LOCAL_STORAGE_KEY = 'piccoloThemeMode'

    var initialMode = localStorage.getItem(LOCAL_STORAGE_KEY)

    if (initialMode) {
        // Make sure the value in local storage is valid
        if (['light', 'dark', 'darkest'].indexOf(initialMode) == -1) {
            initialMode = 'light'
            localStorage.setItem(LOCAL_STORAGE_KEY, initialMode)
        }
    } else {
        // Check if the client prefers dark mode
        if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
            initialMode = 'dark'
        } else {
            initialMode = 'light'
        }
        localStorage.setItem(LOCAL_STORAGE_KEY, initialMode)
    }

    document.documentElement.dataset.mode = initialMode

    PetiteVue.createApp({
        'mode': initialMode,
        handleClick() {
            let currentMode = this.mode

            if (currentMode == 'light') {
                this.mode = 'dark'
            } else if (currentMode == 'dark') {
                this.mode = 'darkest'
            } else if (currentMode == 'darkest') {
                this.mode = 'light'
            }

            document.documentElement.dataset.mode = this.mode
            localStorage.setItem(LOCAL_STORAGE_KEY, this.mode)

            console.log(this.mode)
        }
    }).mount('#mode_toggle')
})()
</script>
            <p class="mobile_search_link">
                <a href="../../search.html" title="Search">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 65 64" fill-rule="evenodd" stroke-linejoin="round" stroke-miterlimit="2">
                        <path d="M14.873 40.009c-2.315-3.943-3.642-8.532-3.642-13.429C11.231 11.91 23.141 0 37.811 0s26.58 11.91 26.58 26.58-11.91 26.58-26.58 26.58a26.44 26.44 0 0 1-14.277-4.161L9.739 62.794a3.12 3.12 0 0 1-4.413 0L.913 58.382c-1.217-1.218-1.217-3.196 0-4.413l13.96-13.96zM37.811 8.054c10.225 0 18.526 8.301 18.526 18.526s-8.301 18.526-18.526 18.526-18.526-8.301-18.526-18.526S27.586 8.054 37.811 8.054z" fill="#fff" />
                    </svg>
                </a>
            </p>
        

        <div class="searchbox_wrapper">
            
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
    </nav>
</div>

    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../guide/creating_a_model.html">1. Creating a Dinf model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/testing_a_model.html">2. Testing a Dinf model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/abc.html">3. An ABC analysis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Topic guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../guide/accuracy.html">Improving discriminator accuracy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/features.html">Feature matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/performance.html">Performance characteristics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/multiple_demes.html">Models with multiple demes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/empirical_data.html">Empirical data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/abc-gan.html">ABC-GAN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/mcmc-gan.html">MCMC-GAN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide/pg-gan.html">PG-GAN (simulated annealing)</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">CLI reference</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Miscellaneous</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../CHANGELOG.html">Changelog</a></li>
</ul>

        </div>
      </div>


    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for dinf.vcf</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">cyvcf2</span>


<div class="viewcode-block" id="get_samples_from_1kgp_metadata"><a class="viewcode-back" href="../../api.html#dinf.get_samples_from_1kgp_metadata">[docs]</a><span class="k">def</span> <span class="nf">get_samples_from_1kgp_metadata</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">populations</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get sample IDs for 1000 Genomes Project populations.</span>
<span class="sd">    Related individuals are removed based on the FatherID and MotherID.</span>

<span class="sd">    :param filename:</span>
<span class="sd">        Path to the file &quot;20130606_g1k_3202_samples_ped_population.txt&quot;</span>
<span class="sd">        This file can be downloaded from</span>
<span class="sd">        http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/</span>
<span class="sd">    :param populations:</span>
<span class="sd">        List of populations to extract.</span>
<span class="sd">    :return:</span>
<span class="sd">        A dict mapping population names to a list of sample IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">recfromtxt</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>
    <span class="c1"># Remove related individuals.</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">FatherID</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">]</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">MotherID</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">data</span><span class="o">.</span><span class="n">SampleID</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">Population</span> <span class="o">==</span> <span class="n">pop</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">populations</span><span class="p">}</span></div>


<div class="viewcode-block" id="get_contig_lengths"><a class="viewcode-back" href="../../api.html#dinf.get_contig_lengths">[docs]</a><span class="k">def</span> <span class="nf">get_contig_lengths</span><span class="p">(</span>
    <span class="n">filename</span><span class="p">:</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span> <span class="o">/</span><span class="p">,</span> <span class="n">keep_contigs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load contig lengths from a whitespace-separated file like an fai (fasta index).</span>

<span class="sd">    The file must have (at least) two columns, the first specifies the</span>
<span class="sd">    contig ID, and the second is the contig length. Additional columns</span>
<span class="sd">    are ignored.</span>

<span class="sd">    :param filename: The path to the file.</span>
<span class="sd">    :param keep_contigs: Use only the specified contigs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">keep_contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">keep_contigs</span><span class="p">)</span>
        <span class="n">seen_contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">maxsplit</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span>
            <span class="n">contig</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seen_contigs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">contig</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">contig</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">keep_contigs</span><span class="p">:</span>
                    <span class="k">continue</span>
            <span class="n">length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">data</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span> <span class="o">=</span> <span class="n">length</span>
    <span class="k">if</span> <span class="n">keep_contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">contigs_not_found</span> <span class="o">=</span> <span class="n">keep_contigs</span> <span class="o">-</span> <span class="n">seen_contigs</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Requested contigs not found: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span></div>


<span class="k">def</span> <span class="nf">filter_missingness</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">max_missing_genotypes</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="n">max_missing_genotypes</span>


<span class="k">def</span> <span class="nf">filter_invariant</span><span class="p">(</span><span class="n">G</span><span class="p">):</span>
    <span class="n">num_non_missing</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac0</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac1</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac2</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="n">ac3</span> <span class="o">=</span> <span class="p">(</span><span class="n">G</span> <span class="o">==</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac0</span><span class="p">,</span> <span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac1</span><span class="p">),</span>
        <span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac2</span><span class="p">,</span> <span class="n">num_non_missing</span> <span class="o">!=</span> <span class="n">ac3</span><span class="p">),</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">get_genotype_matrix</span><span class="p">(</span>
    <span class="n">vcf</span><span class="p">:</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">,</span>
    <span class="o">*</span><span class="p">,</span>
    <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">require_phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a genotype matrix for the specified genomic window.</span>

<span class="sd">    The input region &#39;chrom:start-end&#39; uses 1-based closed coordinates,</span>
<span class="sd">    to match htslib (upon which cyvcf2 is built).</span>
<span class="sd">    The returned matrix has shape (num_sites, num_individuals, ploidy),</span>
<span class="sd">    where each site is a segregating SNP.</span>

<span class="sd">    :param vcf:</span>
<span class="sd">        A cyvcf2.VCF instance from which the genotypes will be obtained.</span>
<span class="sd">    :param chrom:</span>
<span class="sd">        Chromosome name (or non-chromosome sequence name).</span>
<span class="sd">    :param start:</span>
<span class="sd">        1-based start coordinate of the genomic window.</span>
<span class="sd">    :param end:</span>
<span class="sd">        1-based end coordinate of the genomic window.</span>
<span class="sd">    :param max_missing_genotypes:</span>
<span class="sd">        Only retain sites with at most this many missing genotype calls.</span>
<span class="sd">    :param require_phased:</span>
<span class="sd">        If True, raise an error if genotypes are not phased.</span>
<span class="sd">    :return:</span>
<span class="sd">        A 2-tuple of (G, positions) where</span>
<span class="sd">         - ``G`` is a (num_sites, num_individuals, ploidy) genotype matrix,</span>
<span class="sd">         - ``positions`` are the site coordinates, as a zero-based offset from</span>
<span class="sd">           the ``start`` coordinate.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gt_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">positions_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">):</span>
        <span class="c1"># XXX: Many variant fields are broken for ploidy &gt; 2.</span>
        <span class="c1"># https://github.com/brentp/cyvcf2/issues/227</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">variant</span><span class="o">.</span><span class="n">is_snp</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">a</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="n">a</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1"># Check phasing. For ploidy == 1, genotypes are reported as unphased.</span>
        <span class="k">if</span> <span class="n">require_phased</span> <span class="ow">and</span> <span class="n">gt</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="n">a</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unphased genotypes at </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="si">}</span><span class="s2">.</span><span class="se">\n</span><span class="si">{</span><span class="n">variant</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">gt_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gt</span><span class="p">)</span>
        <span class="n">positions_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">POS</span><span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gt_list</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="c1"># Most likely an &quot;inhomogeneous shape&quot; error, caused by different</span>
        <span class="c1"># ploidies at different sites. Catch and reraise to provide</span>
        <span class="c1"># a more user-friendly error message.</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Mismatched ploidy among sites in </span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>

    <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">positions_list</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span> <span class="o">-</span> <span class="n">start</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Filter for missingness.</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">filter_missingness</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">max_missing_genotypes</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">G</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Filter invariant sites.</span>
        <span class="n">keep</span> <span class="o">=</span> <span class="n">filter_invariant</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
        <span class="n">G</span> <span class="o">=</span> <span class="n">G</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">G</span><span class="p">,</span> <span class="n">positions</span>


<div class="viewcode-block" id="BagOfVcf"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf">[docs]</a><span class="k">class</span> <span class="nc">BagOfVcf</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A collection of indexed VCF or BCF files.</span>

<span class="sd">    VCF data are sometimes contained in a single file, and sometimes split into</span>
<span class="sd">    multiple files by chromosome. To remove the burden of dealing with both</span>
<span class="sd">    of these common cases, this class maps a contig ID to a :class:`cyvcf2.VCF`</span>
<span class="sd">    object for that contig. The class implements Python&#39;s :term:`python:mapping`</span>
<span class="sd">    protocol, plus methods for sampling regions of the genome at random.</span>

<span class="sd">    .. code::</span>

<span class="sd">        import dinf</span>

<span class="sd">        vcfs = dinf.BagOfVcf([&quot;chr1.bcf&quot;, &quot;chr2.bcf&quot;, &quot;chr3.bcf&quot;])</span>

<span class="sd">        # Iterate over contig IDs.</span>
<span class="sd">        assert len(vcfs) == 3</span>
<span class="sd">        assert list(vcfs) == [&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr3&quot;]</span>

<span class="sd">        # Lookup a cyvcf2.VCF object by contig ID.</span>
<span class="sd">        chr1 = vcfs[&quot;chr1&quot;]</span>
<span class="sd">        assert chr1.fname == &quot;chr1.bcf&quot;</span>
<span class="sd">        first_variant = next(chr1)</span>
<span class="sd">        assert first_variant.CHROM == &quot;chr1&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">lengths</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Lengths of the contigs in the bag.</span>

<span class="sd">    The order matches the contig order obtained when iterating.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">samples</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">|</span> <span class="kc">None</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A dictionary that maps a label to a list of individual names.</span>

<span class="sd">    The individual names correspond to the VCF columns for which genotypes</span>
<span class="sd">    will be sampled. This is a bookkeeping device that records which genotypes</span>
<span class="sd">    belong to which label (e.g. which population). If None, it is assumed</span>
<span class="sd">    that all individuals in the VCF will be treated as exchangeable.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">files</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="o">/</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">samples</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">contig_lengths</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param files:</span>
<span class="sd">            An iterable of filenames. Each file must be an indexed VCF or BCF,</span>
<span class="sd">            and must contain the FORMAT/GT field.</span>
<span class="sd">        :param contig_lengths:</span>
<span class="sd">            A dict mapping a contig name to contig length. Only the contigs in</span>
<span class="sd">            this dict will be used.</span>
<span class="sd">        :param samples:</span>
<span class="sd">            A dictionary that maps a label to a list of individual names,</span>
<span class="sd">            where the individual names correspond to the VCF columns</span>
<span class="sd">            for which genotypes will be sampled.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">individuals</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="n">samples</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="c1"># Silence some warnings from cyvcf2.</span>
        <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>
            <span class="c1"># We check if a contig is usable for a given vcf by querying</span>
            <span class="c1"># that contig for variants. If there are no variants (e.g. vcfs are</span>
            <span class="c1"># split by chromosome so each vcf has data for only one contig),</span>
            <span class="c1"># then cyvcf2 warns us.</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
                <span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">message</span><span class="o">=</span><span class="s2">&quot;no intervals found&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span>
            <span class="p">)</span>
            <span class="c1"># We detect this condition and raise our own error below.</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span>
                <span class="s2">&quot;ignore&quot;</span><span class="p">,</span>
                <span class="n">message</span><span class="o">=</span><span class="s2">&quot;not all requested samples found&quot;</span><span class="p">,</span>
                <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_fill_bag</span><span class="p">(</span>
                <span class="n">files</span><span class="o">=</span><span class="n">files</span><span class="p">,</span> <span class="n">contig_lengths</span><span class="o">=</span><span class="n">contig_lengths</span><span class="p">,</span> <span class="n">individuals</span><span class="o">=</span><span class="n">individuals</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="n">samples</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">_fill_bag</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">files</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="n">contig_lengths</span><span class="p">:</span> <span class="n">collections</span><span class="o">.</span><span class="n">abc</span><span class="o">.</span><span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">individuals</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a mapping from contig ID to :class:`cyvcf2.VCF` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">contig2file</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">contig2vcf</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">files</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">files</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;File list contains duplicates.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">individuals</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Individuals list contains duplicates.&quot;</span><span class="p">)</span>

        <span class="n">all_samples</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">all_samples_file</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">contig_lengths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">contig_lengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">contigs</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">contigs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">contig_lengths</span><span class="p">)</span>
        <span class="n">contigs_seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">vcf</span> <span class="o">=</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">individuals</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;GT&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">vcf</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> doesn&#39;t contain GT field.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
                <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.tbi&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
                <span class="ow">or</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.csi&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span>
            <span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No index found for </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">all_samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">all_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
                    <span class="n">all_samples_file</span> <span class="o">=</span> <span class="n">file</span>
                <span class="k">elif</span> <span class="n">all_samples</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> has different samples than </span><span class="si">{</span><span class="n">all_samples_file</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">individuals_not_found</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">samples</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Requested individuals not found in </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2">: &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">individuals_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>
            <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">vcf</span><span class="o">.</span><span class="n">seqlens</span>
                <span class="k">except</span> <span class="ne">AttributeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> doesn&#39;t contain contig lengths. &quot;</span>
                        <span class="s2">&quot;You must provide a contig_lengths argument.&quot;</span>
                    <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">vcf</span><span class="o">.</span><span class="n">seqnames</span><span class="p">):</span>
                <span class="n">contigs_seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">contig_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">contig_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="c1"># Check there&#39;s at least one variant. If present, we assume</span>
                <span class="c1"># there exist usable variants (i.e. SNPs) for the contig.</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="nb">next</span><span class="p">(</span><span class="n">vcf</span><span class="p">(</span><span class="n">contig_id</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="n">contig2file</span><span class="p">:</span>
                    <span class="n">first_file</span> <span class="o">=</span> <span class="n">contig2file</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Both </span><span class="si">{</span><span class="n">file</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">first_file</span><span class="si">}</span><span class="s2"> contain records for &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;sequence &#39;</span><span class="si">{</span><span class="n">contig_id</span><span class="si">}</span><span class="s2">&#39;.&quot;</span>
                    <span class="p">)</span>
                <span class="n">contig2file</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">file</span>
                <span class="n">contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcf</span>
                <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">contig_lengths</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcf</span><span class="o">.</span><span class="n">seqlens</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>  <span class="c1"># type: ignore[index]</span>

        <span class="k">if</span> <span class="n">contigs</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">contigs_not_found</span> <span class="o">=</span> <span class="n">contigs</span> <span class="o">-</span> <span class="n">contigs_seen</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Requested contigs not found: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">contigs_not_found</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contig2file</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No usable vcf/bcf files in the list.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">individuals</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">all_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="n">individuals</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_samples</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_individuals</span> <span class="o">=</span> <span class="n">individuals</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span> <span class="o">=</span> <span class="n">contig2file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span> <span class="o">=</span> <span class="n">contig2vcf</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromiter</span><span class="p">(</span><span class="n">contig_lengths</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

        <span class="c1"># The VCF objects are not valid in child processes, so we set a flag</span>
        <span class="c1"># before forking to ensure they&#39;re reopened when used in the child.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">os</span><span class="o">.</span><span class="n">register_at_fork</span><span class="p">(</span><span class="n">before</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_close</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Iterate over contig IDs in the bag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">yield from</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the :class:`cyvcf2.VCF` containing the given contig.</span>

<span class="sd">        :param contig_id:</span>
<span class="sd">            The contig ID.</span>
<span class="sd">        :return:</span>
<span class="sd">            The :class:`cyvcf2.VCF` object.</span>
<span class="sd">        :rtype: cyvcf2.VCF</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">contig_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;key must be a string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_reopen</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The number of contigs in the mapping.</span>

<span class="sd">        :return:</span>
<span class="sd">            The size of the bag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">contig_id</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span><span class="p">[</span><span class="n">contig_id</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">_reopen</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">file2vcf</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">file</span><span class="p">:</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_individuals</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig2vcf</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">contig_id</span><span class="p">:</span> <span class="n">file2vcf</span><span class="p">[</span><span class="n">file</span><span class="p">]</span> <span class="k">for</span> <span class="n">contig_id</span><span class="p">,</span> <span class="n">file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig2file</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_needs_reopen</span> <span class="o">=</span> <span class="kc">False</span>

<div class="viewcode-block" id="BagOfVcf.sample_regions"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf.sample_regions">[docs]</a>    <span class="k">def</span> <span class="nf">sample_regions</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sample a list of (chrom, start, end) triplets.</span>

<span class="sd">        :param size:</span>
<span class="sd">            Number of genomic windows to sample.</span>
<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the sequence to sample.</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            The numpy random number generator.</span>
<span class="sd">        :return:</span>
<span class="sd">            List of 3-tuples: (chrom, start, end).</span>
<span class="sd">            The start and end coordinates are 1-based and inclusive, to match</span>
<span class="sd">            the usual convention for &#39;chrom:start-end&#39;, e.g. in bcftools.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">long_enough</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">&gt;=</span> <span class="n">sequence_length</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">long_enough</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No contigs with length &gt;= </span><span class="si">{</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
        <span class="n">contig_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)[</span><span class="n">long_enough</span><span class="p">]</span>
        <span class="n">contig_lengths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">long_enough</span><span class="p">]</span>

        <span class="n">p</span> <span class="o">=</span> <span class="n">contig_lengths</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">contig_lengths</span><span class="p">)</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">contig_ids</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">p</span><span class="p">)</span>
        <span class="n">upper_limit</span> <span class="o">=</span> <span class="p">(</span><span class="n">contig_lengths</span> <span class="o">-</span> <span class="n">sequence_length</span><span class="p">)[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">rng</span><span class="o">.</span><span class="n">integers</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="n">upper_limit</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">size</span><span class="p">,</span> <span class="n">endpoint</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">sequence_length</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">contig_ids</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="n">regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">regions</span></div>

<div class="viewcode-block" id="BagOfVcf.sample_genotype_matrix"><a class="viewcode-back" href="../../api.html#dinf.BagOfVcf.sample_genotype_matrix">[docs]</a>    <span class="k">def</span> <span class="nf">sample_genotype_matrix</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">min_seg_sites</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">max_missing_genotypes</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">require_phased</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">rng</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">Generator</span><span class="p">,</span>
        <span class="n">retries</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sample a genotype matrix uniformly at random from the genome.</span>

<span class="sd">        :param sequence_length:</span>
<span class="sd">            Length of the sequence to sample.</span>
<span class="sd">        :param max_missing_genotypes:</span>
<span class="sd">            Consider only sites with at most this many missing genotype calls.</span>
<span class="sd">        :param min_seg_sites:</span>
<span class="sd">            Sampled genotype matrix must have at least this many variable</span>
<span class="sd">            sites (after filtering sites for missingness).</span>
<span class="sd">        :param require_phased:</span>
<span class="sd">            If True, raise an error if genotypes are not phased.</span>
<span class="sd">        :param numpy.random.Generator rng:</span>
<span class="sd">            Numpy random number generator.</span>
<span class="sd">        :param retries:</span>
<span class="sd">            Maximum number of attempts allowed to find a genomic window with</span>
<span class="sd">            the required number of segregating sites.</span>
<span class="sd">        :return:</span>
<span class="sd">            A 2-tuple of (G, positions) where</span>
<span class="sd">             - ``G`` is a (num_sites, num_individuals, ploidy) genotype matrix,</span>
<span class="sd">             - ``positions`` are the site coordinates, as an offset from the</span>
<span class="sd">               the start of the genomic window.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">retries</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">retries</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_regions</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">,</span> <span class="n">rng</span><span class="p">)</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_regions</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="n">G</span><span class="p">,</span> <span class="n">positions</span> <span class="o">=</span> <span class="n">get_genotype_matrix</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">[</span><span class="n">chrom</span><span class="p">],</span>
                <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
                <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
                <span class="n">max_missing_genotypes</span><span class="o">=</span><span class="n">max_missing_genotypes</span><span class="p">,</span>
                <span class="n">require_phased</span><span class="o">=</span><span class="n">require_phased</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_seg_sites</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Failed to sample genotype matrix with at least </span><span class="si">{</span><span class="n">min_seg_sites</span><span class="si">}</span><span class="s2"> &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;segregating sites after </span><span class="si">{</span><span class="n">retries</span><span class="si">}</span><span class="s2"> attempts.&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">G</span><span class="p">,</span> <span class="n">positions</span></div></div>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
    

      <div class="clearer"></div>
    </div>
    <div class="button_nav_wrapper">
        <div class="button_nav">
            <div class="left">
                
            </div>

            <div class="right">
                
            </div>
        </div>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2021-2023.
    </div>

<p id="theme_credit">Styled using the <a href="https://github.com/piccolo-orm/piccolo_theme">Piccolo Theme</a></p>
  </body>
</html>